---
title: "Jerry Project"
author: "Xinyi Lin"
date: "4/28/2021"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message=FALSE, warning=FALSE)
```

```{r,message=FALSE,warning=FALSE}
library(splatter)
library(Seurat)
library(SeuratWrappers)
library(speckle)
library(DCATS)
library(ggplot2)
library(tidyverse)
library(tidymodels)
library(diffcyt)
library(scdney)
library(Matrix)
```

```{r}
source("functionsV2.r")
```

## real-world data 1 - Experiment 7

The 'group' column started with 'B' is the indicators of replicates

```{r}
exp7_data <- read.csv("./data/real_world/Exp7_Adam_2017/celltypelabels_Haber.txt", sep="") %>% 
  rownames_to_column("cell") %>% 
  separate(cell, c("batch", "barcode", "condition", "clusterRes"), sep = "_") 

head(exp7_data)
summary(exp7_data)

exp7_data %>% 
  group_by(batch, condition) %>% 
  summarise(n = n())
```

```{r}
geneExpM <- read.delim("./data/real_world/Exp7_Adam_2017/GSE92332_SalmHelm_UMIcounts.txt", header=TRUE)
```

```{r, message=FALSE, warning=FALSE}
seuratObj <- CreateSeuratObject(counts = geneExpM, project="Exp7")
seuratObj <- NormalizeData(seuratObj)
seuratObj <- FindVariableFeatures(seuratObj, selection.method = "vst", nfeatures = 2000)
seuratObj <- ScaleData(seuratObj)
seuratObj <- RunPCA(seuratObj, npcs = 30, verbose = FALSE)
```

```{r}
exp7DF = seuratObj@reductions$pca@cell.embeddings %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("cell") %>% 
  separate(cell, c("batch", "barcode", "condition", "clusterRes"), sep = "_") %>% 
  dplyr::select(-barcode)

## check whether there is duplicate ID -> No with cellID, Yes with only barcode
#exp7 %>% group_by(cell) %>% summarise(n = n()) %>% filter(n != 1)
```

### control vs Hpoly.Day3

```{r}
## data convert
numb_cond1 = exp7_data %>% 
  filter(condition == "Control") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
numb_cond2 = exp7_data %>% 
  filter(condition == "Hpoly.Day3") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
exp7_C3 = exp7_data %>% 
  filter(condition == "Control" | condition == "Hpoly.Day3") %>% mutate(clusterRes = as.factor(clusterRes))
```

Get matrices

```{r}
set.seed(123)
svm_mat = get_similarity_matRW(method = "svm", df = exp7DF)
betabin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100, binom_only = FALSE)
```

```{r}
## results
all_res = data.frame(cluster = rownames(betabin_wMsvm), truth = c(rep("N", 7), "P")) %>% 
  mutate(betabin_wMsvm_pvals = betabin_wMsvm$pvals) %>% 
  mutate(betabin_wMsvm_Res = ifelse(betabin_wMsvm_pvals < 0.05, "P","N"))
all_res
```

```{r}
Info = seuratObj@assays$RNA %>% colnames() %>% str_split("_") %>% as.data.frame()
seuratObj[['group.ident']] = as.character(Info[3,])
Idents(seuratObj) = as.character(Info[4,])
seuratObj$group.ident %>% as.factor() %>% summary()
seuratL = SplitObject(seuratObj, split.by = "group.ident")
```

```{r, eval=FALSE}
## Truth Result
subSeurat = merge(seuratL$Control, seuratL$Hpoly.Day3, add.cell.ids = c("Control", "Hpoly.Day3"))
noDC.markers <- FindMarkers(subSeurat, ident.1 = c("Tuft"), min.pct = 0.25)
noDC.markers
#head(noDC.markers, n = 5)
```

```{r, eval=FALSE}
subSeurat = merge(seuratL$Control, seuratL$Hpoly.Day3, add.cell.ids = c("Control", "Hpoly.Day3"))
noDC.markers <- FindMarkers(subSeurat, ident.1 = c("Endocrine", "Tuft"), min.pct = 0.25)
noDC.markers %>% filter(p_val_adj < 0.05)
```

### control vs Hpoly.Day10

```{r, eval=FALSE}
## data convert
numb_cond1 = exp7_data %>% 
  filter(condition == "Control") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
numb_cond2 = exp7_data %>% 
  filter(condition == "Hpoly.Day10") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
exp7_C10 = exp7_data %>% 
  filter(condition == "Control" | condition == "Hpoly.Day10") %>% mutate(clusterRes = as.factor(clusterRes))
```

```{r, eval=FALSE}
## results
betabin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100, binom_only = FALSE)
all_res = data.frame(cluster = rownames(betabin_wMsvm), truth = c("N","P","N","P","N","N","P","P")) %>% 
  mutate(betabin_wMsvm_pvals = betabin_wMsvm$pvals) %>% 
  mutate(betabin_wMsvm_Res = ifelse(betabin_wMsvm_pvals < 0.05, "P","N"))
all_res
```

```{r, eval=FALSE}
subSeurat = merge(seuratL$Control, seuratL$Hpoly.Day10, add.cell.ids = c("Control", "Hpoly.Day10"))
noDC.markers <- FindMarkers(subSeurat, ident.1 = c("Enterocyte", "Goblet", "TA.Early", "Tuft"), min.pct = 0.25)
noDC.markers %>% filter(p_val_adj < 0.05)
```

### control vs Salmonella

```{r, eval=FALSE}
## data convert
numb_cond1 = exp7_data %>% 
  filter(condition == "Control") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
numb_cond2 = exp7_data %>% 
  filter(condition == "Salmonella") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
exp7_CS = exp7_data %>% 
  filter(condition == "Control" | condition == "Salmonella") %>% mutate(clusterRes = as.factor(clusterRes))
```

```{r, eval=FALSE}
## results
betabin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100, binom_only = FALSE)
all_res = data.frame(cluster = rownames(betabin_wMsvm), truth = c("N","P","N","N","P","P","P","N")) %>% 
  mutate(betabin_wMsvm_pvals = betabin_wMsvm$pvals) %>% 
  mutate(betabin_wMsvm_Res = ifelse(betabin_wMsvm_pvals < 0.05, "P","N"))
all_res
```

```{r, eval=FALSE}
subSeurat = merge(seuratL$Control, seuratL$Salmonella, add.cell.ids = c("Control", "Salmonella"))
noDC.markers <- FindMarkers(subSeurat, ident.1 = c("Enterocyte", "Stem", "TA.Early", "TA"), min.pct = 0.25)
noDC.markers %>% filter(p_val_adj < 0.05)
```

## real-world data 2: Tikhonova_2019

Paper name: Single cell transcriptome profiling of the bone marrow niche at steady state and under stress conditions

```{r}
rw3_clusterInfo = read.csv("./data/real_world/Tikhonova_2019/GSE108891_metadata.csv")
pathL = list.files(path = "./data/real_world/Tikhonova_2019/raw_data", full.names = TRUE, recursive = FALSE)
seuratObjL = vector(mode = "list")
for (i in 1:length(pathL)) {
  rw3_sub = read.table(pathL[i])
  seuratObj <- CreateSeuratObject(counts = rw3_sub, project="rw3")
  seuratObj <- NormalizeData(seuratObj, verbose = FALSE)
  seuratObj <- FindVariableFeatures(seuratObj, selection.method = "vst", nfeatures = 2000, verbose = FALSE)
  seuratObjL[[str_c("sub", as.character(i))]] = seuratObj
}
```

```{r}
anchors <- FindIntegrationAnchors(object.list = seuratObjL, dims = 1:30)
seurat_intg <- IntegrateData(anchorset = anchors, dims = 1:30)
seurat_intg <- ScaleData(seurat_intg)
seurat_intg <- RunPCA(seurat_intg, npcs = 30, verbose = FALSE)
```

```{r}
rw3DF = seurat_intg@reductions$pca@cell.embeddings %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("cell") %>% 
  mutate(cell = str_replace(cell, "X5FU.", "5FU-"),
         cell = str_replace(cell, "CTRL.", "CTRL-"),
         cell = str_replace(cell, "\\.", ":")) %>% 
  merge(rw3_clusterInfo, by = "cell") %>% 
  rename(condition = treatment) %>% 
  mutate(clusterRes = ifelse(cluster %in% c("V1", "V2"), "V", cluster),
         clusterRes = ifelse(clusterRes %in% c("O1", "O2", "O3"), "O", clusterRes)) %>% 
  dplyr::select(-cell, -sample_name, -cluster_long, -tSNE_1, -tSNE_2)
```

```{r}
## data convert
numb_cond1 = rw3DF %>% 
  filter(condition == "CTRL") %>% 
  group_by(clusterRes) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  mutate(P5 = 0) %>%     # there is no P5 in the original paper, add P5
  select(C:P4, P5, V)
numb_cond2 = rw3DF %>% 
  filter(condition == "5FU") %>% 
  group_by(clusterRes) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n")
rw3DF = rw3DF %>% 
  mutate(clusterRes = as.factor(clusterRes))
```

```{r}
set.seed(123)
#svm_mat = get_similarity_matRW(method = "svm", df = rw3DF)
#betabin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100, binom_only = FALSE)
simil_matU = get_similarity_matRW(dim(numb_cond1)[2], confuse_rate=0.1)
betabin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100, binom_only = FALSE)
all_res = data.frame(cluster = rownames(betabin_wMU), truth = c("N","P","N","P","N","N","P","P")) %>% 
  mutate(betabin_wMU_pvals = betabin_wMU$pvals) %>% 
  mutate(betabin_wMU_Res = ifelse(betabin_wMU_pvals < 0.05, "P","N"))
all_res
```

```{r}
Info = seurat_intg@reductions$pca@cell.embeddings %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("ID") %>% 
  mutate(cell = str_replace(ID, "X5FU.", "5FU-"),
         cell = str_replace(cell, "CTRL.", "CTRL-"),
         cell = str_replace(cell, "\\.", ":")) %>% 
  plyr::join(rw3_clusterInfo, by = "cell") %>% 
  mutate(clusterRes = ifelse(cluster %in% c("V1", "V2"), "V", cluster),
         clusterRes = ifelse(clusterRes %in% c("O1", "O2", "O3"), "O", clusterRes)) %>% 
  drop_na()
subSeurat = subset(seurat_intg, cells = Info$ID)
Idents(subSeurat) = Info$clusterRes
```

```{r}
## Truth
noDC.markers <- FindMarkers(subSeurat, ident.1 = c("O", "P2", "P5", "V"), min.pct = 0.25)
noDC.markers
```

## real-world data 3: Baryawno_2019

paper name: A Cellular Taxonomy of the Bone Marrow Stroma in Homeostasis and Leukemia

Read data

```{r}
rw2_geneExpM <- read.delim("./data/real_world/Baryawno_2019/stroma.leuk.ctrl.TP4K.txt", row.names=1)
rw2_clusterInfo = read.delim("./data/real_world/Baryawno_2019/stroma.leuk.ctrl.meta.txt") %>% rename(Cell = NAME)
```

```{r}
seuratObj <- CreateSeuratObject(counts = rw2_geneExpM, project="Bary")
seuratObj <- NormalizeData(seuratObj)
seuratObj <- FindVariableFeatures(seuratObj, selection.method = "vst", nfeatures = 2000)
seuratObj <- ScaleData(seuratObj)
seuratObj <- RunPCA(seuratObj, npcs = 30, verbose = FALSE)
```

filter out cluster 10 since there is only 1 observation of cluster 10. Filter out cluster 10 and cluster 14 since they are not in the original results

```{r}
rw2DF = seuratObj@reductions$pca@cell.embeddings %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("Cell") %>% 
  merge(rw2_clusterInfo, by = "Cell") %>% 
  mutate(Cell = str_replace(Cell, "ctrl_10May", "ctrl1"),
         Cell = str_replace(Cell, "ctrl_16May", "ctrl2"),
         Cell = str_replace(Cell, "ctrl_26May", "ctrl3"),
         Cell = str_replace(Cell, "ctrl_7Jun", "ctrl4"),
         Cell = str_replace(Cell, "ctrl_8May", "ctrl5"),
         Cell = str_replace(Cell, "MLL_10May", "mll1"),
         Cell = str_replace(Cell, "MLL_26May", "mll2"),
         Cell = str_replace(Cell, "MLL_31May", "mll3"),
         Cell = str_replace(Cell, "MLL_8May", "mll4")) %>% 
  rename(cell = Cell, clusterRes = Cluster, condition = Condition) %>% 
  mutate(clusterRes = str_c("cluster", clusterRes)) %>% 
  filter(clusterRes != "cluster10" & clusterRes != "cluster14" ) %>% 
  separate(cell, c("batch", "barcode"), sep = "_") %>% 
  dplyr::select(-barcode)

## check whether there is duplicate ID -> No with cellID, Yes with only barcode
#rw2DF %>% group_by(Cell) %>% summarise(n = n()) %>% filter(n != 1)
```

```{r, eval=FALSE}
rw2DF %>% 
  group_by(batch, clusterRes) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = n)
```

```{r}
## data convert
numb_cond1 = rw2DF %>% 
  filter(condition == "ctrl") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
numb_cond2 = rw2DF %>% 
  filter(condition == "mll") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
rw2DF = rw2DF %>% 
  mutate(clusterRes = as.factor(clusterRes))
numb_cond1[is.na(numb_cond1)] = 0
numb_cond2[is.na(numb_cond2)] = 0
```

Get matrices

```{r}
set.seed(123)
simil_matU = get_similarity_matRW(dim(numb_cond1)[2], confuse_rate=0.1)
betabin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100, binom_only = FALSE)
all_res = data.frame(cluster = rownames(betabin_wMU), truth = c("P","N","N","N","N","P","N","N","P","N","P","P","P","P","P","N")) %>% 
  mutate(betabin_wMU_pvals = betabin_wMU$pvals) %>% 
  mutate(betabin_wMU_Res = ifelse(betabin_wMU_pvals < 0.05, "P","N"))
all_res
```

```{r}
Info = seuratObj@reductions$pca@cell.embeddings %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("Cell") %>% 
  plyr::join(rw2_clusterInfo, by = "Cell") %>% 
  rename(cell = Cell, clusterRes = Cluster, condition = Condition) %>% 
  mutate(clusterRes = str_c("cluster", clusterRes)) %>% 
  .$clusterRes
Idents(seuratObj) = Info
```

```{r}
## Truth
noDC.markers <- FindMarkers(seuratObj, ident.1 = c("cluster0", "cluster15", "cluster2", "cluster4", "cluster5", "cluster6", "cluster7", "cluster8"), min.pct = 0.25)
noDC.markers
```