---
title: "Realworld Data Figures"
output: html_document
---

```{r,message=FALSE,warning=FALSE}
library(splatter)
library(Seurat)
library(SeuratWrappers)
library(speckle)
library(DCATS)
library(ggplot2)
library(tidyverse)
library(tidymodels)
library(diffcyt)
library(scdney)
library(Matrix)
```

```{r}
source("functionsV2.r")
```

## real-world data 1 - Experiment 7

The 'group' column started with 'B' is the indicators of replicates

```{r}
## save data as RData
Haber2017 = list(svmDF = exp7DF, ctrl_Hpoly3 = ctrl_Hpoly3, ctrl_Hpoly10 = ctrl_Hpoly10, ctrl_Salma = ctrl_Salma, source = "A single-cell survey of the small intestinal epithelium")
ctrl_Hpoly3 = list(numb_cond1 = as.matrix(numb_cond1), numb_cond2 = as.matrix(numb_cond2), svm_mat = svm_mat)
ctrl_Hpoly10 = list(numb_cond1 = as.matrix(numb_cond1), numb_cond2 = as.matrix(numb_cond2), svm_mat = svm_mat)
ctrl_Salma = list(numb_cond1 = as.matrix(numb_cond1), numb_cond2 = as.matrix(numb_cond2), svm_mat = svm_mat)
save(Haber2017, file = "./data/real_world/Haber2017.RData")
```

```{r}
exp7_data <- read.csv("./data/real_world/Exp7_Adam_2017/celltypelabels_Haber.txt", sep="") %>% 
  rownames_to_column("cell") %>% 
  separate(cell, c("batch", "barcode", "condition", "clusterRes"), sep = "_") 

head(exp7_data)
summary(exp7_data)

exp7_data %>% 
  group_by(batch, condition) %>% 
  summarise(n = n())
```

```{r}
geneExpM <- read.delim("./data/real_world/Exp7_Adam_2017/GSE92332_SalmHelm_UMIcounts.txt", header=TRUE)
test2 = colnames(geneExpM)
test2a = str_split(test2, "_")
barcode = rep(NA, length(test2))
group = rep(NA, length(test2))
for (i in 1:length(test2)) {
  barcode[i] = test2a[[i]][2]
  group[i] = test2a[[i]][1]
}
test2b = data.frame(barcode = barcode, group = group) %>% 
  filter(barcode %in% exp7_data$barcode)

test2b %>% group_by(group) %>% 
  summarise(n = n())
```

```{r}
seuratObj <- CreateSeuratObject(counts = geneExpM, project="Exp7")
seuratObj <- NormalizeData(seuratObj)
seuratObj <- FindVariableFeatures(seuratObj, selection.method = "vst", nfeatures = 2000)
seuratObj <- ScaleData(seuratObj)
seuratObj <- RunPCA(seuratObj, npcs = 30, verbose = FALSE)
```

```{r}
exp7DF = seuratObj@reductions$pca@cell.embeddings %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("cell") %>% 
  separate(cell, c("batch", "barcode", "condition", "clusterRes"), sep = "_") %>% 
  dplyr::select(-barcode)

## check whether there is duplicate ID -> No with cellID, Yes with only barcode
#exp7 %>% group_by(cell) %>% summarise(n = n()) %>% filter(n != 1)
```

control vs Hpoly.Day3

```{r}
## data convert
numb_cond1 = exp7_data %>% 
  filter(condition == "Control") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
numb_cond2 = exp7_data %>% 
  filter(condition == "Hpoly.Day3") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
exp7_C3 = exp7_data %>% 
  filter(condition == "Control" | condition == "Hpoly.Day3") %>% mutate(clusterRes = as.factor(clusterRes))
```

Get matrices

```{r}
set.seed(123)
simil_matU = get_similarity_matRW(dim(numb_cond1)[2], confuse_rate=0.1)
rf_mat = get_similarity_matRW(method = "rf", df = exp7DF)
svm_mat = get_similarity_matRW(method = "svm", df = exp7DF)
```

Test part

```{r}
betabin_noBC = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = FALSE)
betabin_wBC_U = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = TRUE, similarity_mat = simil_matU)
betabin_wBC_rf = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = TRUE, similarity_mat = rf_mat)
betabin_wBC_svm = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = TRUE, similarity_mat = svm_mat)
## DCATS---betabin with similarity matrix (backward-forward)
betabin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100, binom_only = FALSE)
betabin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100, binom_only = FALSE)
betabin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100, binom_only = FALSE)
bin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100)
bin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100)
bin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100)
## Fisher's exact test
fisher_pvals = getFisher(as.matrix(numb_cond1), as.matrix(numb_cond2))
## speckle
speckleRes = propeller(clusters = exp7_C3$clusterRes, sample = exp7_C3$batch, group = exp7_C3$condition) %>% 
  dplyr::rename(cluster = BaselineProp.clusters,
                      speckle_pvals = FDR) %>%
  dplyr::select(cluster, speckle_pvals)
## scDC
cluster_num = dim(numb_cond1)[2]
res_scDC <- scDC_noClustering(cellTypes = exp7_C3$clusterRes, exp7_C3$batch, calCI = TRUE, calCI_method = c("percentile", "BCa", "multinom"),nboot = 1000, verbose = FALSE)
res_GLM <- fitGLM(res_scDC, c(rep("cond1",dim(numb_cond1)[1]*cluster_num),rep("cond2",dim(numb_cond2)[1]*cluster_num)), pairwise = FALSE, fixed_only = TRUE, verbose = FALSE)
scDCRes_temp = summary(res_GLM$pool_res_fixed)
scDCRes = scDCRes_temp[c(cluster_num+1,(dim(scDCRes_temp)[1]-cluster_num+2):dim(scDCRes_temp)[1]),]
```

```{r}
all_res %>% arrange(betabin_wMU_pvals) %>% .$cluster
```

```{r}
## results
all_res = data.frame(cluster = rownames(betabin_noBC), truth = c(rep("N", 7), "P")) %>% 
  mutate(betabin_noBC_pvals = betabin_noBC$pvals,
         betabin_wBCU_pvals = betabin_wBC_U$pvals,
         betabin_wBCrf_pvals = betabin_wBC_rf$pvals,
         betabin_wBCsvm_pvals = betabin_wBC_svm$pvals,
         betabin_wMU_pvals = betabin_wMU$pvals,
         betabin_wMrf_pvals = betabin_wMrf$pvals,
         betabin_wMsvm_pvals = betabin_wMsvm$pvals,
         bin_wMU_pvals = bin_wMU$pvals,
         bin_wMrf_pvals = bin_wMrf$pvals,
         bin_wMsvm_pvals = bin_wMsvm$pvals,
         fisher_pvals = fisher_pvals,
         scDC_pvals = scDCRes$p.value) %>%
  merge(speckleRes, by = "cluster") %>% 
  mutate(betabin_noBC_Res = ifelse(betabin_noBC_pvals < 0.05, "P","N"),
         betabin_wBCU_Res = ifelse(betabin_wBCU_pvals < 0.05, "P","N"),
         betabin_wBCrf_Res = ifelse(betabin_wBCrf_pvals < 0.05, "P","N"),
         betabin_wBCsvm_Res = ifelse(betabin_wBCsvm_pvals < 0.05, "P","N"),
         betabin_wMU_Res = ifelse(betabin_wMU_pvals < 0.05, "P","N"),
         betabin_wMrf_Res = ifelse(betabin_wMrf_pvals < 0.05, "P","N"),
         betabin_wMsvm_Res = ifelse(betabin_wMsvm_pvals < 0.05, "P","N"),
         bin_wMU_Res = ifelse(bin_wMU_pvals < 0.05, "P","N"),
         bin_wMrf_Res = ifelse(bin_wMrf_pvals < 0.05, "P","N"),
         bin_wMsvm_Res = ifelse(bin_wMsvm_pvals < 0.05, "P","N"),
         fisher_Res = ifelse(fisher_pvals < 0.05, "P","N"),
         scDC_Res = ifelse(scDC_pvals < 0.05, "P","N"),
         speckle_Res = ifelse(speckle_pvals < 0.05, "P","N"))

ResDF = all_res %>% 
  dplyr::select(contains("_Res"))
  #dplyr::select(truth, contains("_Res"))

methodL = str_replace(colnames(ResDF), "_Res", "")
sensitivity = rep(NA, length(methodL))
specificity = rep(NA, length(methodL))
mcc = rep(NA, length(methodL))

for (i in 1:length(methodL)) {
  truth = all_res$truth
  pred = ResDF[, i]
  confu_mtr = table(pred, truth)
  TN = 0
  FN = 0
  FP = 0
  TP = 0
  if(dim(confu_mtr)[1] == 2){
    TN = confu_mtr[1,1] 
    FN = confu_mtr[1,2]
    FP = confu_mtr[2,1]
    TP = confu_mtr[2,2]
    } else {
      if (unique(pred) == "N"){
        TN = confu_mtr[1,1] 
        FN = confu_mtr[1,2]
        } else {
          FP = confu_mtr[1,1]
          TP = confu_mtr[1,2]
        }
      }
  truthN = TN + FP
  truthP = FN + TP
  predP = TP + FP
  predN = FN + TN
  sensitivity[i] = TP/(TP+FN)
  specificity[i] = TN/(TN+FP)
  mcc[i] = (TP*TN-FP*FN)/sqrt(predP*truthP*truthN*predN)
}

resSumDF = data.frame(methodL = methodL, mcc = mcc, sensitivity = sensitivity, specificity = specificity) %>% 
  arrange(desc(mcc))
print(resSumDF)
```

```{r}
library(pROC)
methods = colnames(all_res)[3:15]
roc_rose <- plot(roc(all_res$truth, all_res[,3]), legacy.axes = TRUE, print.auc = TRUE, print.auc.y = .73, print.auc.x = .7, col = 1)
graphics::text(0.1, 0.7, paste(methods[1]), col=1)
for (i in 2:13){
  roc_rose <- plot(roc(all_res$truth, all_res[,i+2]), legacy.axes = TRUE, print.auc = TRUE, print.auc.y = .73-(i-1)*0.07, print.auc.x = .7, col = i, add = TRUE)
  graphics::text(0.1, 0.7-(i-1)*0.07, paste(methods[i]), col=i)
}
```

```{r}
all_res %>% 
  pivot_longer(betabin_noBC_pvals:speckle_pvals, names_to = "method", values_to = "p.value") %>% 
  ggplot(aes(x = cluster, y = p.value, color = method)) +
  geom_bar(stat="identity", fill="white", position=position_dodge())+
  geom_hline(yintercept=0.05, linetype="dashed") +
  theme(axis.text.x = element_text(angle = 45))
```

control vs Hpoly.Day10

```{r}
## data convert
numb_cond1 = exp7_data %>% 
  filter(condition == "Control") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
numb_cond2 = exp7_data %>% 
  filter(condition == "Hpoly.Day10") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
exp7_C10 = exp7_data %>% 
  filter(condition == "Control" | condition == "Hpoly.Day10") %>% mutate(clusterRes = as.factor(clusterRes))
```

Test part

```{r}
betabin_noBC = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = FALSE)
betabin_wBC_U = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = TRUE, similarity_mat = simil_matU)
betabin_wBC_rf = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = TRUE, similarity_mat = rf_mat)
betabin_wBC_svm = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = TRUE, similarity_mat = svm_mat)
## DCATS---betabin with similarity matrix (backward-forward)
betabin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100, binom_only = FALSE)
betabin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100, binom_only = FALSE)
betabin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100, binom_only = FALSE)
bin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100)
bin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100)
bin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100)
## Fisher's exact test
fisher_pvals = getFisher(as.matrix(numb_cond1), as.matrix(numb_cond2))
## speckle
speckleRes = propeller(clusters = exp7_C10$clusterRes, sample = exp7_C10$batch, group = exp7_C10$condition) %>% 
  dplyr::rename(cluster = BaselineProp.clusters,
                      speckle_pvals = FDR) %>%
  dplyr::select(cluster, speckle_pvals)
## scDC
cluster_num = dim(numb_cond1)[2]
res_scDC <- scDC_noClustering(cellTypes = exp7_C10$clusterRes, exp7_C10$batch, calCI = TRUE, calCI_method = c("percentile", "BCa", "multinom"),nboot = 1000, verbose = FALSE)
res_GLM <- fitGLM(res_scDC, c(rep("cond1",dim(numb_cond1)[1]*cluster_num),rep("cond2",dim(numb_cond2)[1]*cluster_num)), pairwise = FALSE, fixed_only = TRUE, verbose = FALSE)
scDCRes_temp = summary(res_GLM$pool_res_fixed)
scDCRes = scDCRes_temp[c(cluster_num+1,(dim(scDCRes_temp)[1]-cluster_num+2):dim(scDCRes_temp)[1]),]
```

```{r}
## results
all_res = data.frame(cluster = rownames(betabin_noBC), truth = c("N","P","N","P","N","N","P","P")) %>% 
  mutate(betabin_noBC_pvals = betabin_noBC$pvals,
         betabin_wBCU_pvals = betabin_wBC_U$pvals,
         betabin_wBCrf_pvals = betabin_wBC_rf$pvals,
         betabin_wBCsvm_pvals = betabin_wBC_svm$pvals,
         betabin_wMU_pvals = betabin_wMU$pvals,
         betabin_wMrf_pvals = betabin_wMrf$pvals,
         betabin_wMsvm_pvals = betabin_wMsvm$pvals,
         bin_wMU_pvals = bin_wMU$pvals,
         bin_wMrf_pvals = bin_wMrf$pvals,
         bin_wMsvm_pvals = bin_wMsvm$pvals,
         fisher_pvals = fisher_pvals,
         scDC_pvals = scDCRes$p.value) %>%
  merge(speckleRes, by = "cluster") %>% 
  mutate(betabin_noBC_Res = ifelse(betabin_noBC_pvals < 0.05, "P","N"),
         #betabin_wBCU_Res = ifelse(betabin_wBCU_pvals < 0.05, "P","N"),
         #betabin_wBCrf_Res = ifelse(betabin_wBCrf_pvals < 0.05, "P","N"),
         #betabin_wBCsvm_Res = ifelse(betabin_wBCsvm_pvals < 0.05, "P","N"),
         betabin_wMU_Res = ifelse(betabin_wMU_pvals < 0.05, "P","N"),
         #betabin_wMrf_Res = ifelse(betabin_wMrf_pvals < 0.05, "P","N"),
         #betabin_wMsvm_Res = ifelse(betabin_wMsvm_pvals < 0.05, "P","N"),
         #bin_wMU_Res = ifelse(bin_wMU_pvals < 0.05, "P","N"),
         #bin_wMrf_Res = ifelse(bin_wMrf_pvals < 0.05, "P","N"),
         #bin_wMsvm_Res = ifelse(bin_wMsvm_pvals < 0.05, "P","N"),
         fisher_Res = ifelse(fisher_pvals < 0.05, "P","N"),
         scDC_Res = ifelse(scDC_pvals < 0.05, "P","N"),
         speckle_Res = ifelse(speckle_pvals < 0.05, "P","N"))

ResDF = all_res %>% 
  dplyr::select(contains("_Res"))
  #dplyr::select(truth, contains("_Res"))

methodL = str_replace(colnames(ResDF), "_Res", "")
sensitivity = rep(NA, length(methodL))
specificity = rep(NA, length(methodL))
mcc = rep(NA, length(methodL))

for (i in 1:length(methodL)) {
  truth = all_res$truth
  pred = ResDF[, i]
  confu_mtr = table(pred, truth)
  TN = 0
  FN = 0
  FP = 0
  TP = 0
  if(dim(confu_mtr)[1] == 2){
    TN = confu_mtr[1,1] 
    FN = confu_mtr[1,2]
    FP = confu_mtr[2,1]
    TP = confu_mtr[2,2]
    } else {
      if (unique(pred) == "N"){
        TN = confu_mtr[1,1] 
        FN = confu_mtr[1,2]
        } else {
          FP = confu_mtr[1,1]
          TP = confu_mtr[1,2]
        }
      }
  truthN = TN + FP
  truthP = FN + TP
  predP = TP + FP
  predN = FN + TN
  sensitivity[i] = TP/(TP+FN)
  specificity[i] = TN/(TN+FP)
  mcc[i] = (TP*TN-FP*FN)/sqrt(predP*truthP*truthN*predN)
}

resSumDF = data.frame(methodL = methodL, mcc = mcc, sensitivity = sensitivity, specificity = specificity) %>% 
  arrange(desc(mcc))
print(resSumDF)
```

```{r}
library(pROC)
methods = colnames(simulationDF)[3:13]
roc_rose <- plot(roc(simulationDF$truth, simulationDF[,3]), legacy.axes = TRUE, print.auc = TRUE, print.auc.y = .73, print.auc.x = .7, col = 1)
graphics::text(0.1, 0.7, paste(methods[1]), col=1)
for (i in 2:11){
  roc_rose <- plot(roc(simulationDF$truth, simulationDF[,i+2]), legacy.axes = TRUE, print.auc = TRUE, print.auc.y = .73-(i-1)*0.07, print.auc.x = .7, col = i, add = TRUE)
  graphics::text(0.1, 0.7-(i-1)*0.07, paste(methods[i]), col=i)
}
```

```{r}
all_res %>% 
  pivot_longer(betabin_noBC_pvals:speckle_pvals, names_to = "method", values_to = "p.value") %>% 
  ggplot(aes(x = cluster, y = p.value, color = method)) +
  geom_bar(stat="identity", fill="white", position=position_dodge())+
  geom_hline(yintercept=0.05, linetype="dashed") +
  theme(axis.text.x = element_text(angle = 45))
```

control vs Salmonella

```{r}
## data convert
numb_cond1 = exp7_data %>% 
  filter(condition == "Control") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
numb_cond2 = exp7_data %>% 
  filter(condition == "Salmonella") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
exp7_CS = exp7_data %>% 
  filter(condition == "Control" | condition == "Salmonella") %>% mutate(clusterRes = as.factor(clusterRes))
```

Test part

```{r}
betabin_noBC = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = FALSE)
betabin_wBC_U = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = TRUE, similarity_mat = simil_matU)
betabin_wBC_rf = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = TRUE, similarity_mat = rf_mat)
betabin_wBC_svm = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = TRUE, similarity_mat = svm_mat)
## DCATS---betabin with similarity matrix (backward-forward)
betabin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100, binom_only = FALSE)
betabin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100, binom_only = FALSE)
betabin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100, binom_only = FALSE)
bin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100)
bin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100)
bin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100)
## Fisher's exact test
fisher_pvals = getFisher(as.matrix(numb_cond1), as.matrix(numb_cond2))
## speckle
speckleRes = propeller(clusters = exp7_CS$clusterRes, sample = exp7_CS$batch, group = exp7_CS$condition) %>% 
  dplyr::rename(cluster = BaselineProp.clusters,
                      speckle_pvals = FDR) %>%
  dplyr::select(cluster, speckle_pvals)
## scDC
cluster_num = dim(numb_cond1)[2]
res_scDC <- scDC_noClustering(cellTypes = exp7_CS$clusterRes, exp7_CS$batch, calCI = TRUE, calCI_method = c("percentile", "BCa", "multinom"),nboot = 1000, verbose = FALSE)
res_GLM <- fitGLM(res_scDC, c(rep("cond1",dim(numb_cond1)[1]*cluster_num),rep("cond2",dim(numb_cond2)[1]*cluster_num)), pairwise = FALSE, fixed_only = TRUE, verbose = FALSE)
scDCRes_temp = summary(res_GLM$pool_res_fixed)
scDCRes = scDCRes_temp[c(cluster_num+1,(dim(scDCRes_temp)[1]-cluster_num+2):dim(scDCRes_temp)[1]),]
```

```{r}
## results
all_res = data.frame(cluster = rownames(betabin_noBC), truth = c("N","P","N","N","P","P","P","N")) %>% 
  mutate(betabin_noBC_pvals = betabin_noBC$pvals,
         betabin_wBCU_pvals = betabin_wBC_U$pvals,
         betabin_wBCrf_pvals = betabin_wBC_rf$pvals,
         betabin_wBCsvm_pvals = betabin_wBC_svm$pvals,
         betabin_wMU_pvals = betabin_wMU$pvals,
         betabin_wMrf_pvals = betabin_wMrf$pvals,
         betabin_wMsvm_pvals = betabin_wMsvm$pvals,
         bin_wMU_pvals = bin_wMU$pvals,
         bin_wMrf_pvals = bin_wMrf$pvals,
         bin_wMsvm_pvals = bin_wMsvm$pvals,
         fisher_pvals = fisher_pvals,
         scDC_pvals = scDCRes$p.value) %>%
  merge(speckleRes, by = "cluster") %>% 
  mutate(betabin_noBC_Res = ifelse(betabin_noBC_pvals < 0.05, "P","N"),
         #betabin_wBCU_Res = ifelse(betabin_wBCU_pvals < 0.05, "P","N"),
         #betabin_wBCrf_Res = ifelse(betabin_wBCrf_pvals < 0.05, "P","N"),
         #betabin_wBCsvm_Res = ifelse(betabin_wBCsvm_pvals < 0.05, "P","N"),
         betabin_wMU_Res = ifelse(betabin_wMU_pvals < 0.05, "P","N"),
         #betabin_wMrf_Res = ifelse(betabin_wMrf_pvals < 0.05, "P","N"),
         #betabin_wMsvm_Res = ifelse(betabin_wMsvm_pvals < 0.05, "P","N"),
         #bin_wMU_Res = ifelse(bin_wMU_pvals < 0.05, "P","N"),
         #bin_wMrf_Res = ifelse(bin_wMrf_pvals < 0.05, "P","N"),
         #bin_wMsvm_Res = ifelse(bin_wMsvm_pvals < 0.05, "P","N"),
         fisher_Res = ifelse(fisher_pvals < 0.05, "P","N"),
         scDC_Res = ifelse(scDC_pvals < 0.05, "P","N"),
         speckle_Res = ifelse(speckle_pvals < 0.05, "P","N"))

ResDF = all_res %>% 
  dplyr::select(contains("_Res"))
  #dplyr::select(truth, contains("_Res"))

methodL = str_replace(colnames(ResDF), "_Res", "")
sensitivity = rep(NA, length(methodL))
specificity = rep(NA, length(methodL))
mcc = rep(NA, length(methodL))

for (i in 1:length(methodL)) {
  truth = all_res$truth
  pred = ResDF[, i]
  confu_mtr = table(pred, truth)
  TN = 0
  FN = 0
  FP = 0
  TP = 0
  if(dim(confu_mtr)[1] == 2){
    TN = confu_mtr[1,1] 
    FN = confu_mtr[1,2]
    FP = confu_mtr[2,1]
    TP = confu_mtr[2,2]
    } else {
      if (unique(pred) == "N"){
        TN = confu_mtr[1,1] 
        FN = confu_mtr[1,2]
        } else {
          FP = confu_mtr[1,1]
          TP = confu_mtr[1,2]
        }
      }
  truthN = TN + FP
  truthP = FN + TP
  predP = TP + FP
  predN = FN + TN
  sensitivity[i] = TP/(TP+FN)
  specificity[i] = TN/(TN+FP)
  mcc[i] = (TP*TN-FP*FN)/sqrt(predP*truthP*truthN*predN)
}

resSumDF = data.frame(methodL = methodL, mcc = mcc, sensitivity = sensitivity, specificity = specificity) %>% 
  arrange(desc(mcc))
print(resSumDF)
```

```{r}
all_res %>% 
  pivot_longer(betabin_noBC_pvals:speckle_pvals, names_to = "method", values_to = "p.value") %>% 
  ggplot(aes(x = cluster, y = p.value, color = method)) +
  geom_bar(stat="identity", fill="white", position=position_dodge())+
  geom_hline(yintercept=0.05, linetype="dashed") +
  theme(axis.text.x = element_text(angle = 45))
```

## real-world data 2

Can't find out the cluster results for each cell. Can't use 'svm_mat' and 'rf_mat'.

Read Data

```{r, eval=FALSE}
pathL = list.dirs(path = "./data/real_world/Baryawno_2019", full.names = TRUE, recursive = FALSE)
c = 0
m = 0
seuratObjL = vector(mode = "list")
for (i in 1:length(pathL)) {
  expression_matrix1 <- Read10X(data.dir = pathL[i])
  seuratObj = CreateSeuratObject(counts = expression_matrix1)
  if (str_split(pathL[i], "_")[[1]][4] == "ctrl"){
    c = c+1
    sample = str_c("ctrl", as.character(c))
  }
  if (str_split(pathL[i], "_")[[1]][4] == "MLL"){
    m = m+1
    sample = str_c("mll", as.character(m))
  }
  seuratObj = AddMetaData(object = seuratObj, metadata = rep(sample, dim(seuratObj)[2]), col.name = 'sample')
  seuratObjL[[sample]] = seuratObj
}
```

### Use seurat to integrate data

```{r, eval=FALSE}
for (i in 1:length(seuratObjL)) {
  seuratObjL[[i]] <- NormalizeData(seuratObjL[[i]], verbose = FALSE)
  seuratObjL[[i]] <- FindVariableFeatures(seuratObjL[[i]], selection.method = "vst", nfeatures = 2000, verbose = FALSE)
}
anchors <- FindIntegrationAnchors(object.list = seuratObjL, dims = 1:30)
seurat_intg <- IntegrateData(anchorset = anchors, dims = 1:30)
save(seurat_intg, file = "./data/real_world/Baryawno_2019/seurat_intg.RData")
```

```{r, eval=FALSE}
load("./data/real_world/Baryawno_2019/seurat_intg.RData")
DefaultAssay(seurat_intg) <- "integrated"
seurat_intg <- ScaleData(seurat_intg, verbose = FALSE)
seurat_intg <- RunPCA(seurat_intg, npcs = 30, verbose = FALSE)
```

Read data

```{r}
rw2_geneExpM <- read.delim("./data/real_world/Baryawno_2019/stroma.leuk.ctrl.TP4K.txt", row.names=1)
rw2_clusterInfo = read.delim("./data/real_world/Baryawno_2019/stroma.leuk.ctrl.meta.txt") %>% rename(Cell = NAME)
```

```{r}
seuratObj <- CreateSeuratObject(counts = rw2_geneExpM, project="Bary")
seuratObj <- NormalizeData(seuratObj)
seuratObj <- FindVariableFeatures(seuratObj, selection.method = "vst", nfeatures = 2000)
seuratObj <- ScaleData(seuratObj)
seuratObj <- RunPCA(seuratObj, npcs = 30, verbose = FALSE)
```

filter out cluster 10 since there is only 1 observation of cluster 10. Filter out cluster 10 and cluster 14 since they are not in the original results

```{r}
rw2DF = seuratObj@reductions$pca@cell.embeddings %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("Cell") %>% 
  merge(rw2_clusterInfo, by = "Cell") %>% 
  mutate(Cell = str_replace(Cell, "ctrl_10May", "ctrl1"),
         Cell = str_replace(Cell, "ctrl_16May", "ctrl2"),
         Cell = str_replace(Cell, "ctrl_26May", "ctrl3"),
         Cell = str_replace(Cell, "ctrl_7Jun", "ctrl4"),
         Cell = str_replace(Cell, "ctrl_8May", "ctrl5"),
         Cell = str_replace(Cell, "MLL_10May", "mll1"),
         Cell = str_replace(Cell, "MLL_26May", "mll2"),
         Cell = str_replace(Cell, "MLL_31May", "mll3"),
         Cell = str_replace(Cell, "MLL_8May", "mll4")) %>% 
  rename(cell = Cell, clusterRes = Cluster, condition = Condition) %>% 
  mutate(clusterRes = str_c("cluster", clusterRes)) %>% 
  filter(clusterRes != "cluster10" & clusterRes != "cluster14" ) %>% 
  separate(cell, c("batch", "barcode"), sep = "_") %>% 
  dplyr::select(-barcode)

## check whether there is duplicate ID -> No with cellID, Yes with only barcode
#rw2DF %>% group_by(Cell) %>% summarise(n = n()) %>% filter(n != 1)
```

```{r}
rw2DF %>% 
  group_by(batch, clusterRes) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = n)
```

```{r}
## data convert
numb_cond1 = rw2DF %>% 
  filter(condition == "ctrl") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
numb_cond2 = rw2DF %>% 
  filter(condition == "mll") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
rw2DF = rw2DF %>% 
  mutate(clusterRes = as.factor(clusterRes))
```

```{r}
print(numb_cond1)
print(numb_cond2)
```

Get matrices

```{r}
set.seed(123)
simil_matU = get_similarity_matRW(dim(numb_cond1)[2], confuse_rate=0.1)
rf_mat = get_similarity_matRW(method = "rf", df = rw2DF)
svm_mat = get_similarity_matRW(method = "svm", df = rw2DF)
```

Test part(svm predition doesn't contain cluster 17 and dcats can't work in this situation, so not using it)

```{r}
numb_cond1[is.na(numb_cond1)] = 0
numb_cond2[is.na(numb_cond2)] = 0
betabin_noBC = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = FALSE)
betabin_wBC_U = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = TRUE, similarity_mat = simil_matU)
betabin_wBC_rf = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = TRUE, similarity_mat = rf_mat)
#betabin_wBC_svm = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = TRUE, similarity_mat = svm_mat)
## DCATS---betabin with similarity matrix (backward-forward)
betabin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100, binom_only = FALSE)
betabin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100, binom_only = FALSE)
#betabin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100, binom_only = FALSE)
bin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100)
bin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100)
#bin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100)
## Fisher's exact test
fisher_pvals = getFisher(as.matrix(numb_cond1), as.matrix(numb_cond2))
## speckle
speckleRes = propeller(clusters = rw2DF$clusterRes, sample = rw2DF$batch, group = rw2DF$condition) %>% 
  dplyr::rename(cluster = BaselineProp.clusters,
                      speckle_pvals = FDR) %>%
  dplyr::select(cluster, speckle_pvals)
## scDC
cluster_num = dim(numb_cond1)[2]
res_scDC <- scDC_noClustering(cellTypes = rw2DF$clusterRes, rw2DF$batch, calCI = TRUE, calCI_method = c("percentile", "BCa", "multinom"),nboot = 1000, verbose = FALSE)
res_GLM <- fitGLM(res_scDC, c(rep("cond1",dim(numb_cond1)[1]*cluster_num),rep("cond2",dim(numb_cond2)[1]*cluster_num)), pairwise = FALSE, fixed_only = TRUE, verbose = FALSE)
scDCRes_temp = summary(res_GLM$pool_res_fixed)
scDCRes = scDCRes_temp[c(cluster_num+1,(dim(scDCRes_temp)[1]-cluster_num+2):dim(scDCRes_temp)[1]),]
```

```{r}
## results
all_res = data.frame(cluster = rownames(betabin_noBC), truth = c("P","N","N","N","N","P","N","N","P","N","P","P","P","P","P","N")) %>%
  mutate(betabin_noBC_pvals = betabin_noBC$pvals,
         betabin_wBCU_pvals = betabin_wBC_U$pvals,
         betabin_wBCrf_pvals = betabin_wBC_rf$pvals,
         #betabin_wBCsvm_pvals = betabin_wBC_svm$pvals,
         betabin_wMU_pvals = betabin_wMU$pvals,
         betabin_wMrf_pvals = betabin_wMrf$pvals,
         #betabin_wMsvm_pvals = betabin_wMsvm$pvals,
         bin_wMU_pvals = bin_wMU$pvals,
         bin_wMrf_pvals = bin_wMrf$pvals,
         #bin_wMsvm_pvals = bin_wMsvm$pvals,
         fisher_pvals = fisher_pvals,
         scDC_pvals = scDCRes$p.value) %>%
  merge(speckleRes, by = "cluster") %>% 
  mutate(truth = ifelse(truth == "P", 1, 0),
         betabin_noBC_Res = ifelse(betabin_noBC_pvals < 0.05, "P","N"),
         betabin_wBCU_Res = ifelse(betabin_wBCU_pvals < 0.05, "P","N"),
         betabin_wBCrf_Res = ifelse(betabin_wBCrf_pvals < 0.05, "P","N"),
         #betabin_wBCsvm_Res = ifelse(betabin_wBCsvm_pvals < 0.05, "P","N"),
         betabin_wMU_Res = ifelse(betabin_wMU_pvals < 0.05, "P","N"),
         betabin_wMrf_Res = ifelse(betabin_wMrf_pvals < 0.05, "P","N"),
         #betabin_wMsvm_Res = ifelse(betabin_wMsvm_pvals < 0.05, "P","N"),
         bin_wMU_Res = ifelse(bin_wMU_pvals < 0.05, "P","N"),
         bin_wMrf_Res = ifelse(bin_wMrf_pvals < 0.05, "P","N"),
         #bin_wMsvm_Res = ifelse(bin_wMsvm_pvals < 0.05, "P","N"),
         fisher_Res = ifelse(fisher_pvals < 0.05, "P","N"),
         scDC_Res = ifelse(scDC_pvals < 0.05, "P","N"),
         speckle_Res = ifelse(speckle_pvals < 0.05, "P","N"))

ResDF = all_res %>% 
  dplyr::select(contains("_Res"))

methodL = str_replace(colnames(ResDF), "_Res", "")
sensitivity = rep(NA, length(methodL))
specificity = rep(NA, length(methodL))

for (i in 1:length(methodL)) {
  truth = all_res$truth
  pred = ResDF[, i]
  confu_mtr = table(pred, truth)
  TN = 0
  FN = 0
  FP = 0
  TP = 0
  if(dim(confu_mtr)[1] == 2){
    TN = confu_mtr[1,1] 
    FN = confu_mtr[1,2]
    FP = confu_mtr[2,1]
    TP = confu_mtr[2,2]
    } else {
      if (unique(pred) == "N"){
        TN = confu_mtr[1,1] 
        FN = confu_mtr[1,2]
        } else {
          FP = confu_mtr[1,1]
          TP = confu_mtr[1,2]
        }
      }
  truthN = TN + FP
  truthP = FN + TP
  predP = TP + FP
  predN = FN + TN
  sensitivity[i] = TP/(TP+FN)
  specificity[i] = TN/(TN+FP)
}

resSumDF = data.frame(methodL = methodL, sensitivity = sensitivity, specificity = specificity)
```

```{r}
all_res %>% 
  pivot_longer(betabin_noBC_pvals:speckle_pvals, names_to = "method", values_to = "p.value") %>% 
  ggplot(aes(x = cluster, y = p.value, color = method)) +
  geom_bar(stat="identity", fill="white", position=position_dodge())+
  geom_hline(yintercept=0.05, linetype="dashed") +
  theme(axis.text.x = element_text(angle = 45))
```

## real-world data 3

```{r}
rw3_clusterInfo = read.csv("./data/real_world/Tikhonova_2019/GSE108891_metadata.csv")
rw3_clusterInfo %>% 
  separate(cell, c("trail", "bardcode"), sep = ":") %>% 
  group_by(trail) %>% 
  summarise(n = n())
rw3_clusterInfo %>% 
  group_by(cluster) %>% 
  summarise(n = n())
```

```{r}
pathL = list.files(path = "./data/real_world/Tikhonova_2019/raw_data", full.names = TRUE, recursive = FALSE)
seuratObjL = vector(mode = "list")
for (i in 1:length(pathL)) {
  rw3_sub = read.table(pathL[i])
  seuratObj <- CreateSeuratObject(counts = rw3_sub, project="rw3")
  seuratObj <- NormalizeData(seuratObj, verbose = FALSE)
  seuratObj <- FindVariableFeatures(seuratObj, selection.method = "vst", nfeatures = 2000, verbose = FALSE)
  seuratObjL[[str_c("sub", as.character(i))]] = seuratObj
}
```

```{r}
anchors <- FindIntegrationAnchors(object.list = seuratObjL, dims = 1:30)
seurat_intg <- IntegrateData(anchorset = anchors, dims = 1:30)
seurat_intg <- ScaleData(seurat_intg)
seurat_intg <- RunPCA(seurat_intg, npcs = 30, verbose = FALSE)
```

```{r}
rw3DF = seurat_intg@reductions$pca@cell.embeddings %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("cell") %>% 
  mutate(cell = str_replace(cell, "X5FU.", "5FU-"),
         cell = str_replace(cell, "CTRL.", "CTRL-"),
         cell = str_replace(cell, "\\.", ":")) %>% 
  merge(rw3_clusterInfo, by = "cell") %>% 
  rename(condition = treatment) %>% 
  mutate(clusterRes = ifelse(cluster %in% c("V1", "V2"), "V", cluster),
         clusterRes = ifelse(clusterRes %in% c("O1", "O2", "O3"), "O", clusterRes)) %>% 
  dplyr::select(-cell, -sample_name, -cluster_long, -tSNE_1, -tSNE_2)
```

```{r}
## data convert
numb_cond1 = rw3DF %>% 
  filter(condition == "CTRL") %>% 
  group_by(clusterRes) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  mutate(P5 = 0) %>%     # there is no P5 in the original paper, add P5
  select(C:P4, P5, V)
numb_cond2 = rw3DF %>% 
  filter(condition == "5FU") %>% 
  group_by(clusterRes) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n")
rw3DF = rw3DF %>% 
  mutate(clusterRes = as.factor(clusterRes))
```

```{r}
print(numb_cond1)
print(numb_cond2)
```

Get matrices

```{r}
set.seed(123)
simil_matU = get_similarity_matRW(dim(numb_cond1)[2], confuse_rate=0.1)
rf_mat = get_similarity_matRW(method = "rf", df = rw2DF)
svm_mat = get_similarity_matRW(method = "svm", df = rw2DF)
```

```{r}
betabin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100, binom_only = FALSE)
betabin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100, binom_only = FALSE)
betabin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100, binom_only = FALSE)
bin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100)
bin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100)
bin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100)
## Fisher's exact test
fisher_pvals = getFisher(as.matrix(numb_cond1), as.matrix(numb_cond2))
```

```{r}
all_res = data.frame(cluster = rownames(betabin_wMU), truth = c("N","P","N","P","N","N","P","P")) %>%
  mutate(betabin_wMU_pvals = betabin_wMU$pvals,
         betabin_wMrf_pvals = betabin_wMrf$pvals,
         betabin_wMsvm_pvals = betabin_wMsvm$pvals,
         bin_wMU_pvals = bin_wMU$pvals,
         bin_wMrf_pvals = bin_wMrf$pvals,
         bin_wMsvm_pvals = bin_wMsvm$pvals,
         fisher_pvals = fisher_pvals) %>%
  mutate(truth = ifelse(truth == "P", 1, 0),
         betabin_wMU_Res = ifelse(betabin_wMU_pvals < 0.05, "P","N"),
         betabin_wMrf_Res = ifelse(betabin_wMrf_pvals < 0.05, "P","N"),
         betabin_wMsvm_Res = ifelse(betabin_wMsvm_pvals < 0.05, "P","N"),
         bin_wMU_Res = ifelse(bin_wMU_pvals < 0.05, "P","N"),
         bin_wMrf_Res = ifelse(bin_wMrf_pvals < 0.05, "P","N"),
         bin_wMsvm_Res = ifelse(bin_wMsvm_pvals < 0.05, "P","N"),
         fisher_Res = ifelse(fisher_pvals < 0.05, "P","N"))

ResDF = all_res %>% 
  dplyr::select(contains("_Res"))

methodL = str_replace(colnames(ResDF), "_Res", "")
sensitivity = rep(NA, length(methodL))
specificity = rep(NA, length(methodL))

for (i in 1:length(methodL)) {
  truth = all_res$truth
  pred = ResDF[, i]
  confu_mtr = table(pred, truth)
  TN = 0
  FN = 0
  FP = 0
  TP = 0
  if(dim(confu_mtr)[1] == 2){
    TN = confu_mtr[1,1] 
    FN = confu_mtr[1,2]
    FP = confu_mtr[2,1]
    TP = confu_mtr[2,2]
    } else {
      if (unique(pred) == "N"){
        TN = confu_mtr[1,1] 
        FN = confu_mtr[1,2]
        } else {
          FP = confu_mtr[1,1]
          TP = confu_mtr[1,2]
        }
      }
  truthN = TN + FP
  truthP = FN + TP
  predP = TP + FP
  predN = FN + TN
  sensitivity[i] = TP/(TP+FN)
  specificity[i] = TN/(TN+FP)
}

resSumDF = data.frame(methodL = methodL, sensitivity = sensitivity, specificity = specificity)
```

## real-world data 4: Exp3

```{r}
exp3_clusterInfo <- read.delim("./data/real_world/Exp3_Farbehi_2019/TIP_tSNE_cluster_ID_table.txt") %>% 
  mutate(cluster = str_replace(cluster, "Î¦", "phi")) %>% 
  filter(cluster %in% c("M1Mphi", "EC1", "F-SL", "F-SH", "BC", "M2Mphi", "M1Mo", "MYO", "DC", "Mural", "MAC-IFNIC", "MAC8"))

exp3_clusterInfo %>% 
  group_by(experiment) %>% 
  summarise(n = n())

exp3_clusterInfo %>% 
  group_by(experiment, cluster) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = cluster, values_from = n) %>% 
  column_to_rownames("experiment")
```

```{r}
exp3_data <- read.csv("./data/real_world/Exp3_Farbehi_2019/TIP_ShamVsMI_days3_7.txt", sep="") %>% 
  as.matrix()
print(exp3_data[1:6, 1:6])

exp3_geneExpM = exp3_data[,exp3_clusterInfo$cell_barcode]
```

```{r}
seuratObj <- CreateSeuratObject(counts = exp3_geneExpM, project="Exp3")
seuratObj <- NormalizeData(seuratObj)
seuratObj <- FindVariableFeatures(seuratObj, selection.method = "vst", nfeatures = 2000)
seuratObj <- ScaleData(seuratObj)
seuratObj <- RunPCA(seuratObj, npcs = 30, verbose = FALSE)
```

```{r}
exp3DF = seuratObj@reductions$pca@cell.embeddings %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("cell_barcode") %>% 
  merge(exp3_clusterInfo) %>% 
  rename(clusterRes = cluster, condition = experiment) %>% 
  mutate(condition = str_replace(condition, " ", "")) %>% 
  dplyr::select(-tSNE_1, -tSNE_2, -cell_barcode)

head(exp3DF)
```

Sham vs MI-Day 3

```{r}
## data convert
numb_cond1 = exp3DF %>% 
  filter(condition == "Sham") %>% 
  group_by(clusterRes) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n")
  
numb_cond2 = exp3DF %>% 
  filter(condition == "MI-Day3") %>% 
  group_by(clusterRes) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n")

exp3DF = exp3DF %>% 
  mutate(clusterRes = as.factor(clusterRes))
```

```{r}
print(numb_cond1)
print(numb_cond2)
```

Get matrices

```{r}
set.seed(123)
simil_matU = get_similarity_matRW(dim(numb_cond1)[2], confuse_rate=0.1)
rf_mat = get_similarity_matRW(method = "rf", df = exp3DF)
svm_mat = get_similarity_matRW(method = "svm", df = exp3DF)
```

```{r}
set.seed(123)
betabin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100, binom_only = FALSE)
betabin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100, binom_only = FALSE)
betabin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100, binom_only = FALSE)
bin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100)
bin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100)
bin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100)
## Fisher's exact test
fisher_pvals = getFisher(as.matrix(numb_cond1), as.matrix(numb_cond2))
```

```{r}
all_res = data.frame(cluster = rownames(betabin_wMU), truth = c("P","P","P","P","P","P","P","N","P","P","P","N")) %>%
  mutate(#betabin_wMU_pvals = betabin_wMU$pvals,
         #betabin_wMrf_pvals = betabin_wMrf$pvals,
         #betabin_wMsvm_pvals = betabin_wMsvm$pvals,
         bin_wMU_pvals = bin_wMU$pvals,
         bin_wMrf_pvals = bin_wMrf$pvals,
         bin_wMsvm_pvals = bin_wMsvm$pvals,
         fisher_pvals = fisher_pvals) %>%
  mutate(#betabin_wMU_Res = ifelse(betabin_wMU_pvals < 0.05, "P","N"),
         #betabin_wMrf_Res = ifelse(betabin_wMrf_pvals < 0.05, "P","N"),
         #betabin_wMsvm_Res = ifelse(betabin_wMsvm_pvals < 0.05, "P","N"),
         bin_wMU_Res = ifelse(bin_wMU_pvals < 0.05, "P","N"),
         bin_wMrf_Res = ifelse(bin_wMrf_pvals < 0.05, "P","N"),
         bin_wMsvm_Res = ifelse(bin_wMsvm_pvals < 0.05, "P","N"),
         fisher_Res = ifelse(fisher_pvals < 0.05, "P","N"))

ResDF = all_res %>% 
  dplyr::select(contains("_Res"))

methodL = str_replace(colnames(ResDF), "_Res", "")
sensitivity = rep(NA, length(methodL))
specificity = rep(NA, length(methodL))

for (i in 1:length(methodL)) {
  truth = all_res$truth
  pred = ResDF[, i]
  confu_mtr = table(pred, truth)
  TN = 0
  FN = 0
  FP = 0
  TP = 0
  if(dim(confu_mtr)[1] == 2){
    TN = confu_mtr[1,1] 
    FN = confu_mtr[1,2]
    FP = confu_mtr[2,1]
    TP = confu_mtr[2,2]
    } else {
      if (unique(pred) == "N"){
        TN = confu_mtr[1,1] 
        FN = confu_mtr[1,2]
        } else {
          FP = confu_mtr[1,1]
          TP = confu_mtr[1,2]
        }
      }
  truthN = TN + FP
  truthP = FN + TP
  predP = TP + FP
  predN = FN + TN
  sensitivity[i] = TP/(TP+FN)
  specificity[i] = TN/(TN+FP)
}

resSumDF = data.frame(methodL = methodL, sensitivity = sensitivity, specificity = specificity)
```

Sham vs MI-Day 7

```{r}
## data convert
numb_cond1 = exp3DF %>% 
  filter(condition == "Sham") %>% 
  group_by(clusterRes) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n")
  
numb_cond2 = exp3DF %>% 
  filter(condition == "MI-Day7") %>% 
  group_by(clusterRes) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n")

exp3DF = exp3DF %>% 
  mutate(clusterRes = as.factor(clusterRes))
```

```{r}
print(numb_cond1)
print(numb_cond2)
```

```{r}
set.seed(12)
betabin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100, binom_only = FALSE)
betabin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100, binom_only = FALSE)
betabin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100, binom_only = FALSE)
bin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100)
bin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100)
bin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100)
## Fisher's exact test
fisher_pvals = getFisher(as.matrix(numb_cond1), as.matrix(numb_cond2))
```

```{r}
all_res = data.frame(cluster = rownames(betabin_wMU), truth = c("N","N","P","P","P","N","N","P","N","N","N","P")) %>%
  mutate(#betabin_wMU_pvals = betabin_wMU$pvals,
         #betabin_wMrf_pvals = betabin_wMrf$pvals,
         #betabin_wMsvm_pvals = betabin_wMsvm$pvals,
         bin_wMU_pvals = bin_wMU$pvals,
         bin_wMrf_pvals = bin_wMrf$pvals,
         bin_wMsvm_pvals = bin_wMsvm$pvals,
         fisher_pvals = fisher_pvals) %>%
  mutate(#betabin_wMU_Res = ifelse(betabin_wMU_pvals < 0.05, "P","N"),
         #betabin_wMrf_Res = ifelse(betabin_wMrf_pvals < 0.05, "P","N"),
         #betabin_wMsvm_Res = ifelse(betabin_wMsvm_pvals < 0.05, "P","N"),
         bin_wMU_Res = ifelse(bin_wMU_pvals < 0.05, "P","N"),
         bin_wMrf_Res = ifelse(bin_wMrf_pvals < 0.05, "P","N"),
         bin_wMsvm_Res = ifelse(bin_wMsvm_pvals < 0.05, "P","N"),
         fisher_Res = ifelse(fisher_pvals < 0.05, "P","N"))

ResDF = all_res %>% 
  dplyr::select(contains("_Res"),truth, cluster)

methodL = str_replace(colnames(ResDF), "_Res", "")
sensitivity = rep(NA, length(methodL))
specificity = rep(NA, length(methodL))

for (i in 1:length(methodL)) {
  truth = all_res$truth
  pred = ResDF[, i]
  confu_mtr = table(pred, truth)
  TN = 0
  FN = 0
  FP = 0
  TP = 0
  if(dim(confu_mtr)[1] == 2){
    TN = confu_mtr[1,1] 
    FN = confu_mtr[1,2]
    FP = confu_mtr[2,1]
    TP = confu_mtr[2,2]
    } else {
      if (unique(pred) == "N"){
        TN = confu_mtr[1,1] 
        FN = confu_mtr[1,2]
        } else {
          FP = confu_mtr[1,1]
          TP = confu_mtr[1,2]
        }
      }
  truthN = TN + FP
  truthP = FN + TP
  predP = TP + FP
  predN = FN + TN
  sensitivity[i] = TP/(TP+FN)
  specificity[i] = TN/(TN+FP)
}

resSumDF = data.frame(methodL = methodL, sensitivity = sensitivity, specificity = specificity)
```

MI-Day 3 vs MI-Day 7

```{r}
## data convert
numb_cond1 = exp3DF %>% 
  filter(condition == "MI-Day3") %>% 
  group_by(clusterRes) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n")
  
numb_cond2 = exp3DF %>% 
  filter(condition == "MI-Day7") %>% 
  group_by(clusterRes) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n")

exp3DF = exp3DF %>% 
  mutate(clusterRes = as.factor(clusterRes))
```

```{r}
print(numb_cond1)
print(numb_cond2)
```

```{r}
set.seed(12)
betabin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100, binom_only = FALSE)
betabin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100, binom_only = FALSE)
betabin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100, binom_only = FALSE)
bin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100)
bin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100)
bin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100)
## Fisher's exact test
fisher_pvals = getFisher(as.matrix(numb_cond1), as.matrix(numb_cond2))
```

```{r}
all_res = data.frame(cluster = rownames(betabin_wMU), truth = c("P","P","P","N","N","P","P","P","N","N","N","P")) %>%
  mutate(#betabin_wMU_pvals = betabin_wMU$pvals,
         #betabin_wMrf_pvals = betabin_wMrf$pvals,
         #betabin_wMsvm_pvals = betabin_wMsvm$pvals,
         bin_wMU_pvals = bin_wMU$pvals,
         bin_wMrf_pvals = bin_wMrf$pvals,
         bin_wMsvm_pvals = bin_wMsvm$pvals,
         fisher_pvals = fisher_pvals) %>%
  mutate(#betabin_wMU_Res = ifelse(betabin_wMU_pvals < 0.05, "P","N"),
         #betabin_wMrf_Res = ifelse(betabin_wMrf_pvals < 0.05, "P","N"),
         #betabin_wMsvm_Res = ifelse(betabin_wMsvm_pvals < 0.05, "P","N"),
         bin_wMU_Res = ifelse(bin_wMU_pvals < 0.05, "P","N"),
         bin_wMrf_Res = ifelse(bin_wMrf_pvals < 0.05, "P","N"),
         bin_wMsvm_Res = ifelse(bin_wMsvm_pvals < 0.05, "P","N"),
         fisher_Res = ifelse(fisher_pvals < 0.05, "P","N"))

ResDF = all_res %>% 
  dplyr::select(contains("_Res"))

methodL = str_replace(colnames(ResDF), "_Res", "")
sensitivity = rep(NA, length(methodL))
specificity = rep(NA, length(methodL))

for (i in 1:length(methodL)) {
  truth = all_res$truth
  pred = ResDF[, i]
  confu_mtr = table(pred, truth)
  TN = 0
  FN = 0
  FP = 0
  TP = 0
  if(dim(confu_mtr)[1] == 2){
    TN = confu_mtr[1,1] 
    FN = confu_mtr[1,2]
    FP = confu_mtr[2,1]
    TP = confu_mtr[2,2]
    } else {
      if (unique(pred) == "N"){
        TN = confu_mtr[1,1] 
        FN = confu_mtr[1,2]
        } else {
          FP = confu_mtr[1,1]
          TP = confu_mtr[1,2]
        }
      }
  truthN = TN + FP
  truthP = FN + TP
  predP = TP + FP
  predN = FN + TN
  sensitivity[i] = TP/(TP+FN)
  specificity[i] = TN/(TN+FP)
}

resSumDF = data.frame(methodL = methodL, sensitivity = sensitivity, specificity = specificity)
```

## real-world data 5: Exp6-demuxlet

```{r}
## save data as demonstarate data
Kang2017 = list(numb_cond1 = as.matrix(numb_cond1), numb_cond2 = as.matrix(numb_cond2), svm_mat = svm_mat, svmDF = exp6DF, source = "Multiplexed droplet single-cell RNA-sequencing using natural genetic variation")
save(Kang2017, file = "./data/real_world/Kang2017.RData")
```

```{r}
graphs = seurat_intg@graphs$RNA_snn
labels = exp6DF$clusterRes
knn_mat = KNN_transition(graphs, labels)
order1 = colnames(numb_cond1)
simil_matK = knn_mat[order1, order1]
```

### one replicate version

```{r}
exp6_clusterInfo = read.delim("./data/real_world/Exp6_Kang/GSE96583_batch2.total.tsne.df.tsv") %>% 
  dplyr::rename(clusterRes = cell, condition = stim) %>% 
  rownames_to_column("cell") %>% 
  filter(!is.na(clusterRes))
head(exp6_clusterInfo)
exp6_clusterInfo %>% group_by(condition, clusterRes) %>% 
  summarise(n = n())
## check whether exist duplicate barcode -> no
exp6_clusterInfo %>% 
  group_by(cell) %>% 
  summarise(n = n()) %>% 
  filter(n > 1)
```

```{r}
matrix1 = readMM(gzfile("./data/real_world/Exp6_Kang/GSM2560248_2.1.mtx.gz")) %>% 
  as.matrix()
matrix2 = readMM(gzfile("./data/real_world/Exp6_Kang/GSM2560249_2.2.mtx.gz")) %>% 
  as.matrix()
```

```{r}
data_dir1 = "./data/real_world/Exp6_Kang/GSM2560248"
list.files(data_dir1) # Should show barcodes.tsv, genes.tsv, and matrix.mtx
expression_matrix1 <- Read10X(data.dir = data_dir1)
seuratObj1 = CreateSeuratObject(counts = expression_matrix1)
seuratObj1 = AddMetaData(object = seuratObj1, metadata = rep("ctrl", dim(seuratObj1)[2]), col.name = 'sample')

data_dir2 = "./data/real_world/Exp6_Kang/GSM2560249"
list.files(data_dir2) # Should show barcodes.tsv, genes.tsv, and matrix.mtx
expression_matrix2 <- Read10X(data.dir = data_dir2)
seuratObj2 = CreateSeuratObject(counts = expression_matrix2)
seuratObj2 = AddMetaData(object = seuratObj2, metadata = rep("stim", dim(seuratObj2)[2]), col.name = 'sample')
```

```{r}
seuratObjL = c(sub1 = seuratObj1, sub2 = seuratObj2)
```

```{r}
anchors <- FindIntegrationAnchors(object.list = seuratObjL, dims = 1:30)
seurat_intg <- IntegrateData(anchorset = anchors, dims = 1:30)
seurat_intg <- ScaleData(seurat_intg)
seurat_intg <- RunPCA(seurat_intg, npcs = 30, verbose = FALSE)
```

```{r}
exp6DF = seurat_intg@reductions$pca@cell.embeddings %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("cell") %>% 
  mutate(cell = str_replace(cell, "_[12]", "")) %>% 
  merge(exp6_clusterInfo) %>% 
  dplyr::select(-cell, -cluster, -multiplets, -tsne1, -tsne2, -ind)

head(exp6DF)
```

```{r}
## data convert
numb_cond1 = exp6DF %>% 
  filter(condition == "ctrl") %>% 
  group_by(clusterRes) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n")
  
numb_cond2 = exp6DF %>% 
  filter(condition == "stim") %>% 
  group_by(clusterRes) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n")

exp6DF = exp6DF %>% 
  mutate(clusterRes = as.factor(clusterRes))
```

```{r}
print(numb_cond1)
print(numb_cond2)
```

Get matrices

```{r}
simil_matU = get_similarity_matRW(dim(numb_cond1)[2], confuse_rate=0.1)
rf_mat = get_similarity_matRW(method = "rf", df = exp6DF)
svm_mat = get_similarity_matRW(method = "svm", df = exp6DF)
```

```{r}
set.seed(123)
betabin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100, binom_only = FALSE)
betabin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100, binom_only = FALSE)
betabin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100, binom_only = FALSE)
bin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100)
bin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100)
bin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100)
## Fisher's exact test
fisher_pvals = getFisher(as.matrix(numb_cond1), as.matrix(numb_cond2))
```

```{r}
all_res = data.frame(cluster = rownames(betabin_wMU), truth = rep("N", 8)) %>%
  mutate(betabin_wMU_pvals = betabin_wMU$pvals,
         betabin_wMrf_pvals = betabin_wMrf$pvals,
         betabin_wMsvm_pvals = betabin_wMsvm$pvals,
         bin_wMU_pvals = bin_wMU$pvals,
         bin_wMrf_pvals = bin_wMrf$pvals,
         bin_wMsvm_pvals = bin_wMsvm$pvals,
         fisher_pvals = fisher_pvals) %>%
  mutate(betabin_wMU_Res = ifelse(betabin_wMU_pvals < 0.05, "P","N"),
         betabin_wMrf_Res = ifelse(betabin_wMrf_pvals < 0.05, "P","N"),
         betabin_wMsvm_Res = ifelse(betabin_wMsvm_pvals < 0.05, "P","N"),
         bin_wMU_Res = ifelse(bin_wMU_pvals < 0.05, "P","N"),
         bin_wMrf_Res = ifelse(bin_wMrf_pvals < 0.05, "P","N"),
         bin_wMsvm_Res = ifelse(bin_wMsvm_pvals < 0.05, "P","N"),
         fisher_Res = ifelse(fisher_pvals < 0.05, "P","N"))

ResDF = all_res %>% 
  dplyr::select(contains("_Res"), truth, cluster)

methodL = str_replace(colnames(ResDF), "_Res", "")
sensitivity = rep(NA, length(methodL))
specificity = rep(NA, length(methodL))

for (i in 1:length(methodL)) {
  truth = all_res$truth
  pred = ResDF[, i]
  confu_mtr = table(pred, truth)
  TN = 0
  FN = 0
  FP = 0
  TP = 0
  if(dim(confu_mtr)[1] == 2){
    TN = confu_mtr[1,1] 
    FN = confu_mtr[1,2]
    FP = confu_mtr[2,1]
    TP = confu_mtr[2,2]
    } else {
      if (unique(pred) == "N"){
        TN = confu_mtr[1,1] 
        FN = confu_mtr[1,2]
        } else {
          FP = confu_mtr[1,1]
          TP = confu_mtr[1,2]
        }
      }
  truthN = TN + FP
  truthP = FN + TP
  predP = TP + FP
  predN = FN + TN
  sensitivity[i] = TP/(TP+FN)
  specificity[i] = TN/(TN+FP)
}

resSumDF = data.frame(methodL = methodL, sensitivity = sensitivity, specificity = specificity)
```



### multiple replicates version

```{r}
ctrl_repInfo <- read.delim("./data/real_world/Exp6_Kang/ye1.ctrl.8.10.sm.best")
stim_repInfo <- read.delim("./data/real_world/Exp6_Kang/ye2.stim.8.10.sm.best")
head(ctrl_repInfo)
head(stim_repInfo)

exp6_repInfo = rbind(ctrl_repInfo, stim_repInfo) %>% 
  dplyr::rename(cell = BARCODE, batch = BEST) %>% 
  filter(str_detect(batch, 'SNG')) %>% 
  dplyr::select(cell, batch)
head(exp6_repInfo)
```

```{r}
exp6DF_mul = seurat_intg@reductions$pca@cell.embeddings %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("cell") %>% 
  mutate(cell = str_replace(cell, "_[12]", "")) %>% 
  merge(exp6_clusterInfo) %>% 
  merge(exp6_repInfo) %>% 
  mutate(batch = str_c(condition, batch)) %>% 
  dplyr::select(-cell, -cluster, -multiplets, -tsne1, -tsne2, -ind)

head(exp6DF_mul)
```

```{r}
## data convert
numb_cond1 = exp6DF_mul %>% 
  filter(condition == "ctrl") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
numb_cond2 = exp6DF_mul %>% 
  filter(condition == "stim") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
```

Test part

```{r}
betabin_noBC = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = FALSE)
betabin_wBC_U = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = TRUE, similarity_mat = simil_matU)
betabin_wBC_rf = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = TRUE, similarity_mat = rf_mat)
betabin_wBC_svm = betabinLRT_rw(as.matrix(numb_cond1), as.matrix(numb_cond2), bias_corr = TRUE, similarity_mat = svm_mat)
## DCATS---betabin with similarity matrix (backward-forward)
betabin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100, binom_only = FALSE)
betabin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100, binom_only = FALSE)
betabin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100, binom_only = FALSE)
bin_wMU = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = simil_matU, n_samples = 100)
bin_wMrf = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = rf_mat, n_samples = 100)
bin_wMsvm = dcats_betabinRW(as.matrix(numb_cond1), as.matrix(numb_cond2), similarity_mat = svm_mat, n_samples = 100)
## Fisher's exact test
fisher_pvals = getFisher(as.matrix(numb_cond1), as.matrix(numb_cond2))
## speckle
speckleRes = speckle::propeller(clusters = exp6DF_mul$clusterRes, sample = exp6DF_mul$batch, group = exp6DF_mul$condition) %>% 
  dplyr::rename(cluster = BaselineProp.clusters,
                      speckle_pvals = FDR) %>%
  dplyr::select(cluster, speckle_pvals)
## scDC
cluster_num = dim(numb_cond1)[2]
res_scDC <- scDC_noClustering(cellTypes = exp6DF_mul$clusterRes, exp6DF_mul$batch, calCI = TRUE, calCI_method = c("percentile", "BCa", "multinom"),nboot = 1000, verbose = FALSE)
res_GLM <- scdney::fitGLM(res_scDC, c(rep("cond1",dim(numb_cond1)[1]*cluster_num),rep("cond2",dim(numb_cond2)[1]*cluster_num)), pairwise = FALSE, fixed_only = TRUE, verbose = FALSE)
scDCRes_temp = summary(res_GLM$pool_res_fixed)
scDCRes = scDCRes_temp[c(cluster_num+1,(dim(scDCRes_temp)[1]-cluster_num+2):dim(scDCRes_temp)[1]),]
```

```{r}
## results
all_res = data.frame(cluster = rownames(betabin_noBC), truth = rep("N", 8)) %>% 
  mutate(betabin_noBC_pvals = betabin_noBC$pvals,
         betabin_wBCU_pvals = betabin_wBC_U$pvals,
         betabin_wBCrf_pvals = betabin_wBC_rf$pvals,
         betabin_wBCsvm_pvals = betabin_wBC_svm$pvals,
         betabin_wMU_pvals = betabin_wMU$pvals,
         betabin_wMrf_pvals = betabin_wMrf$pvals,
         betabin_wMsvm_pvals = betabin_wMsvm$pvals,
         bin_wMU_pvals = bin_wMU$pvals,
         bin_wMrf_pvals = bin_wMrf$pvals,
         bin_wMsvm_pvals = bin_wMsvm$pvals,
         fisher_pvals = fisher_pvals,
         scDC_pvals = scDCRes$p.value) %>%
  merge(speckleRes, by = "cluster") %>% 
  mutate(betabin_noBC_Res = ifelse(betabin_noBC_pvals < 0.05, "P","N"),
         betabin_wBCU_Res = ifelse(betabin_wBCU_pvals < 0.05, "P","N"),
         betabin_wBCrf_Res = ifelse(betabin_wBCrf_pvals < 0.05, "P","N"),
         betabin_wBCsvm_Res = ifelse(betabin_wBCsvm_pvals < 0.05, "P","N"),
         betabin_wMU_Res = ifelse(betabin_wMU_pvals < 0.05, "P","N"),
         betabin_wMrf_Res = ifelse(betabin_wMrf_pvals < 0.05, "P","N"),
         betabin_wMsvm_Res = ifelse(betabin_wMsvm_pvals < 0.05, "P","N"),
         bin_wMU_Res = ifelse(bin_wMU_pvals < 0.05, "P","N"),
         bin_wMrf_Res = ifelse(bin_wMrf_pvals < 0.05, "P","N"),
         bin_wMsvm_Res = ifelse(bin_wMsvm_pvals < 0.05, "P","N"),
         fisher_Res = ifelse(fisher_pvals < 0.05, "P","N"),
         scDC_Res = ifelse(scDC_pvals < 0.05, "P","N"),
         speckle_Res = ifelse(speckle_pvals < 0.05, "P","N"))

ResDF = all_res %>% 
  dplyr::select(contains("_Res"))

methodL = str_replace(colnames(ResDF), "_Res", "")
sensitivity = rep(NA, length(methodL))
specificity = rep(NA, length(methodL))
mcc = rep(NA, length(methodL))

for (i in 1:length(methodL)) {
  truth = all_res$truth
  pred = ResDF[, i]
  confu_mtr = table(pred, truth)
  TN = 0
  FN = 0
  FP = 0
  TP = 0
  if(dim(confu_mtr)[1] == 2){
    TN = confu_mtr[1,1] 
    FN = confu_mtr[1,2]
    FP = confu_mtr[2,1]
    TP = confu_mtr[2,2]
    } else {
      if (unique(pred) == "N"){
        TN = confu_mtr[1,1] 
        FN = confu_mtr[1,2]
        } else {
          FP = confu_mtr[1,1]
          TP = confu_mtr[1,2]
        }
      }
  truthN = TN + FP
  truthP = FN + TP
  predP = TP + FP
  predN = FN + TN
  sensitivity[i] = TP/(TP+FN)
  specificity[i] = TN/(TN+FP)
  mcc[i] = (TP*TN-FP*FN)/sqrt(predP*truthP*truthN*predN)
}

resSumDF = data.frame(methodL = methodL, mcc = mcc, sensitivity = sensitivity, specificity = specificity) %>% 
  arrange(desc(mcc))
print(resSumDF)

all_res %>% 
  dplyr::select(cluster, truth, betabin_noBC_Res, betabin_wMU_Res, fisher_Res, scDC_Res, speckle_Res)
```




