---
title: "Realworld Data Figures"
output: html_document
---

```{r,message=FALSE,warning=FALSE}
library(splatter)
library(Seurat)
library(SeuratWrappers)
library(speckle)
library(DCATS)
library(ggplot2)
library(tidyverse)
library(tidymodels)
library(diffcyt)
library(scDC)
library(Matrix)
library(miloR)
library(SingleCellExperiment)
```

```{r}
source("functionsV2.r")
```


## real-world data 1 - Experiment 7

### codes to generate data we need and store it

The 'group' column started with 'B' is the indicators of replicates

```{r, eval=FALSE}
## save data as RData
Haber2017 = list(svmDF = exp7DF, ctrl_Hpoly3 = ctrl_Hpoly3, ctrl_Hpoly10 = ctrl_Hpoly10, ctrl_Salma = ctrl_Salma, source = "A single-cell survey of the small intestinal epithelium")
ctrl_Hpoly3 = list(numb_cond1 = as.matrix(numb_cond1), numb_cond2 = as.matrix(numb_cond2), svm_mat = svm_mat)
ctrl_Hpoly10 = list(numb_cond1 = as.matrix(numb_cond1), numb_cond2 = as.matrix(numb_cond2), svm_mat = svm_mat)
ctrl_Salma = list(numb_cond1 = as.matrix(numb_cond1), numb_cond2 = as.matrix(numb_cond2), svm_mat = svm_mat)
save(Haber2017, file = "./data/real_world/Haber2017.RData")
```

```{r}
exp7_data <- read.csv("./data/real_world/Exp7_Adam_2017/celltypelabels_Haber.txt", sep="") %>% 
  rownames_to_column("cell") %>% 
  separate(cell, c("batch", "barcode", "condition", "clusterRes"), sep = "_") 

head(exp7_data)
summary(exp7_data)

exp7_data %>% 
  group_by(batch, condition) %>% 
  summarise(n = n())
```

```{r, eval=FALSE}
geneExpM <- read.delim("./data/real_world/Exp7_Adam_2017/GSE92332_SalmHelm_UMIcounts.txt", header=TRUE)
test2 = colnames(geneExpM)
test2a = str_split(test2, "_")
barcode = rep(NA, length(test2))
group = rep(NA, length(test2))
for (i in 1:length(test2)) {
  barcode[i] = test2a[[i]][2]
  group[i] = test2a[[i]][1]
}
test2b = data.frame(barcode = barcode, group = group) %>% 
  filter(barcode %in% exp7_data$barcode)

test2b %>% group_by(group) %>% 
  summarise(n = n())
```

```{r, eval=FALSE}
seuratObj <- CreateSeuratObject(counts = geneExpM, project="Exp7")
seuratObj <- NormalizeData(seuratObj)
seuratObj <- FindVariableFeatures(seuratObj, selection.method = "vst", nfeatures = 2000)
seuratObj <- ScaleData(seuratObj)
seuratObj <- RunPCA(seuratObj, npcs = 30, verbose = FALSE)
```

```{r, eval=FALSE}
exp7DF = seuratObj@reductions$pca@cell.embeddings %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("cell") %>% 
  separate(cell, c("batch", "barcode", "condition", "clusterRes"), sep = "_") %>% 
  dplyr::select(-barcode)
seuratObj[['batch']] = exp7DF$batch
seuratObj[['condition']] = exp7DF$condition
seuratObj[['clusterRes']] = exp7DF$clusterRes
Idents(seuratObj) = 'condition'
## check whether there is duplicate ID -> No with cellID, Yes with only barcode
#exp7 %>% group_by(cell) %>% summarise(n = n()) %>% filter(n != 1)
```

control vs Hpoly.Day3

```{r}
## data convert
numb_cond1 = exp7_data %>% 
  filter(condition == "Control") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
numb_cond2 = exp7_data %>% 
  filter(condition == "Hpoly.Day3") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
exp7_C3 = exp7_data %>% 
  filter(condition == "Control" | condition == "Hpoly.Day3") %>% mutate(clusterRes = as.factor(clusterRes))
```

Get matrices

```{r, eval=FALSE}
set.seed(123)
simil_matU = get_similarity_matRW(dim(Haber2017$count_ctrl)[2], confuse_rate=0.1)
rf_mat = get_similarity_matRW(method = "rf", df = exp7DF)
svm_mat = get_similarity_matRW(method = "svm", df = exp7DF)
```

### Test part

```{r}
data("Haber2017")
countM = rbind(Haber2017$count_ctrl, Haber2017$count_Hpoly3) %>% as.matrix()
designM = data.frame(condition = c(rep("g1", nrow(Haber2017$count_ctrl)), rep("g2", nrow(Haber2017$count_Hpoly3))))
#betabin_null = dcats_GLM(countM, designM)
#wtoPhi_emSVM = dcats_GLM(countM, designM, similarity_mat = Haber2017$svm_mat, n_samples = NULL)
betabin_null = eachPhi(countM, designM)
wtoPhi_emSVM = eachPhi(countM, designM, similarity_mat = Haber2017$svm_mat, n_samples = NULL)
estPhi = getPhi(countM, designM)
estPhi_null = dcats_GLM(countM, designM, fix_phi = estPhi)
estPhi_emSVM = dcats_GLM(countM, designM, similarity_mat = Haber2017$svm_mat, fix_phi = estPhi)
## Fisher's exact test
fisher_pvals = getFisher(as.matrix(Haber2017$count_ctrl), as.matrix(Haber2017$count_Hpoly3))
## speckle
speckleRes = propeller(clusters = exp7_C3$clusterRes, sample = exp7_C3$batch, group = exp7_C3$condition) %>% 
  dplyr::rename(cluster = BaselineProp.clusters,
                      speckle_pvals = FDR) %>%
  dplyr::select(cluster, speckle_pvals)
## scDC
cluster_num = dim(numb_cond1)[2]
res_scDC <- scDC_noClustering(cellTypes = exp7_C3$clusterRes, exp7_C3$batch, calCI = TRUE, calCI_method = c("percentile", "BCa", "multinom"),nboot = 1000, verbose = FALSE)
res_GLM <- fitGLM(res_scDC, c(rep("cond1",dim(numb_cond1)[1]*cluster_num),rep("cond2",dim(numb_cond2)[1]*cluster_num)), pairwise = FALSE, fixed_only = TRUE, verbose = FALSE)
scDCRes_temp = summary(res_GLM$pool_res_fixed)
scDCRes = scDCRes_temp[c(cluster_num+1,(dim(scDCRes_temp)[1]-cluster_num+2):dim(scDCRes_temp)[1]),]
```

```{r}
plot(colSums(Haber2017$count_ctrl)+colSums(Haber2017$count_Hpoly3), betabin_null$fm0_phi)
abline(h = phi, col = 'red')
plot(colSums(Haber2017$count_ctrl)+colSums(Haber2017$count_Hpoly3), betabin_null$fm1_phi)
abline(h = phi, col = 'red')
plot(colSums(Haber2017$count_ctrl)+colSums(Haber2017$count_Hpoly3), wtoPhi_emSVM$fm0_phi)
abline(h = phi, col = 'red')
plot(colSums(Haber2017$count_ctrl)+colSums(Haber2017$count_Hpoly3), wtoPhi_emSVM$fm1_phi)
abline(h = phi, col = 'red')
```

## Milo

```{r}
testObj = subset(x = seuratObj, idents = c("Control", "Hpoly.Day3"))
Idents(testObj) = 'clusterRes'
miloRes = getMilo(testObj)$da_results %>% 
  filter(ident_fraction > 0.8) %>%
    mutate(direct = ifelse( logFC > 0  & SpatialFDR < 0.1, "milo_more", "neutral"),
           direct = ifelse( logFC < 0  & SpatialFDR < 0.1, "milo_less", direct)) %>% 
    group_by(ident, direct) %>% 
    summarise(n = n()) %>% 
    pivot_wider(names_from = direct, values_from = n) %>% 
    replace(is.na(.), 0)%>% 
    dplyr::rename(cluster = ident)
  if ("milo_less" %!in% colnames(miloRes)) {miloRes$milo_less = rep(0, nrow(miloRes))}
  if ("milo_more" %!in% colnames(miloRes)) {miloRes$milo_more = rep(0, nrow(miloRes))}
```

```{r}
## results
all_resC3 = data.frame(cluster = rownames(betabin_null$LRT_pvals), truth = c(rep("N", 7), "P")) %>% 
  mutate(betabin_null_pvals = as.vector(betabin_null$LRT_pvals),
         wtoPhi_emSVM_pvals = as.vector(wtoPhi_emSVM$LRT_pvals),
         estPhi_null_pvals = as.vector(estPhi_null$LRT_pvals),
         estPhi_emSVM_pvals = as.vector(estPhi_emSVM$LRT_pvals),
         fisher_pvals = fisher_pvals,
         scDC_pvals = scDCRes$p.value) %>%
  merge(speckleRes, by = "cluster") %>% 
  mutate(betabin_null_Res = ifelse(betabin_null_pvals < 0.01, "***", "n.s."),
         betabin_null_Res = ifelse(0.01<betabin_null_pvals&betabin_null_pvals<0.05, "**", betabin_null_Res),
         betabin_null_Res = ifelse(0.05<betabin_null_pvals&betabin_null_pvals<0.1, "*", betabin_null_Res)) %>% 
  #dplyr::select(truth, betabin_null_Res, betabin_null_pvals)
  mutate(wtoPhi_emSVM_Res = ifelse(wtoPhi_emSVM_pvals < 0.01, "***", "n.s."),
         wtoPhi_emSVM_Res = ifelse(0.01<wtoPhi_emSVM_pvals&wtoPhi_emSVM_pvals<0.05, "**", wtoPhi_emSVM_Res),
         wtoPhi_emSVM_Res = ifelse(0.05<wtoPhi_emSVM_pvals&wtoPhi_emSVM_pvals<0.1, "*", wtoPhi_emSVM_Res)) %>%
  #dplyr::select(truth, wtoPhi_emSVM_Res, wtoPhi_emSVM_pvals)
  mutate(estPhi_null_Res = ifelse(estPhi_null_pvals < 0.01, "***", "n.s."),
         estPhi_null_Res = ifelse(0.01<estPhi_null_pvals&estPhi_null_pvals<0.05, "**", estPhi_null_Res),
         estPhi_null_Res = ifelse(0.05<estPhi_null_pvals&estPhi_null_pvals<0.1, "*", estPhi_null_Res)) %>%
  mutate(estPhi_emSVM_Res = ifelse(estPhi_emSVM_pvals < 0.01, "***", "n.s."),
         estPhi_emSVM_Res = ifelse(0.01<estPhi_emSVM_pvals&estPhi_emSVM_pvals<0.05, "**", estPhi_emSVM_Res),
         estPhi_emSVM_Res = ifelse(0.05<estPhi_emSVM_pvals&estPhi_emSVM_pvals<0.1, "*", estPhi_emSVM_Res)) %>%
  mutate(fisher_Res = ifelse(fisher_pvals < 0.01, "***", "n.s."),
         fisher_Res = ifelse(0.01<fisher_pvals&fisher_pvals<0.05, "**", fisher_Res),
         fisher_Res = ifelse(0.05<fisher_pvals&fisher_pvals<0.1, "*", fisher_Res)) %>%
  #dplyr::select(truth, fisher_Res, fisher_pvals)
  mutate(scDC_Res = ifelse(scDC_pvals < 0.01, "***", "n.s."),
         scDC_Res = ifelse(0.01<scDC_pvals&scDC_pvals<0.05, "**", scDC_Res),
         scDC_Res = ifelse(0.05<scDC_pvals&scDC_pvals<0.1, "*", scDC_Res)) %>%
  #dplyr::select(truth, scDC_Res, scDC_pvals)
  mutate(speckle_Res = ifelse(speckle_pvals < 0.01, "***", "n.s."),
         speckle_Res = ifelse(0.01<speckle_pvals&speckle_pvals<0.05, "**", speckle_Res),
         speckle_Res = ifelse(0.05<speckle_pvals&speckle_pvals<0.1, "*", speckle_Res)) %>% 
  merge(miloRes, by = "cluster")


all_resC3 %>% 
  dplyr::select(cluster, truth, contains("_Res"))
```

control vs Hpoly.Day10

```{r}
## data convert
numb_cond1 = exp7_data %>% 
  filter(condition == "Control") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
numb_cond2 = exp7_data %>% 
  filter(condition == "Hpoly.Day10") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
exp7_C10 = exp7_data %>% 
  filter(condition == "Control" | condition == "Hpoly.Day10") %>% mutate(clusterRes = as.factor(clusterRes))
```

Test part

```{r}
data("Haber2017")
countM = rbind(Haber2017$count_ctrl, Haber2017$count_Hpoly10) %>% as.matrix()
designM = data.frame(condition = c(rep("g1", nrow(Haber2017$count_ctrl)), rep("g2", nrow(Haber2017$count_Hpoly3))))
#betabin_null = dcats_GLM(countM, designM)
#wtoPhi_emSVM = dcats_GLM(countM, designM, similarity_mat = Haber2017$svm_mat, n_samples = NULL)
betabin_null = eachPhi(countM, designM)
wtoPhi_emSVM = eachPhi(countM, designM, similarity_mat = Haber2017$svm_mat)
estPhi = getPhi(countM, designM)
estPhi_null = dcats_GLM(countM, designM, fix_phi = estPhi)
estPhi_emSVM = dcats_GLM(countM, designM, similarity_mat = Haber2017$svm_mat, fix_phi = estPhi)
## Fisher's exact test
fisher_pvals = getFisher(as.matrix(Haber2017$count_ctrl), as.matrix(Haber2017$count_Hpoly10))
## speckle
speckleRes = propeller(clusters = exp7_C10$clusterRes, sample = exp7_C10$batch, group = exp7_C10$condition) %>% 
  dplyr::rename(cluster = BaselineProp.clusters,
                      speckle_pvals = FDR) %>%
  dplyr::select(cluster, speckle_pvals)
## scDC
cluster_num = dim(numb_cond1)[2]
res_scDC <- scDC_noClustering(cellTypes = exp7_C10$clusterRes, exp7_C10$batch, calCI = TRUE, calCI_method = c("percentile", "BCa", "multinom"),nboot = 1000, verbose = FALSE)
res_GLM <- fitGLM(res_scDC, c(rep("cond1",dim(numb_cond1)[1]*cluster_num),rep("cond2",dim(numb_cond2)[1]*cluster_num)), pairwise = FALSE, fixed_only = TRUE, verbose = FALSE)
scDCRes_temp = summary(res_GLM$pool_res_fixed)
scDCRes = scDCRes_temp[c(cluster_num+1,(dim(scDCRes_temp)[1]-cluster_num+2):dim(scDCRes_temp)[1]),]
```

```{r}
plot(colSums(Haber2017$count_ctrl)+colSums(Haber2017$count_Hpoly10), betabin_null$fm0_phi)
abline(h = phi, col = 'red')
plot(colSums(Haber2017$count_ctrl)+colSums(Haber2017$count_Hpoly10), betabin_null$fm1_phi)
abline(h = phi, col = 'red')
plot(colSums(Haber2017$count_ctrl)+colSums(Haber2017$count_Hpoly10), wtoPhi_emSVM$fm0_phi)
abline(h = phi, col = 'red')
plot(colSums(Haber2017$count_ctrl)+colSums(Haber2017$count_Hpoly10), wtoPhi_emSVM$fm1_phi)
abline(h = phi, col = 'red')
```

```{r}
set.seed(123)
testObj = subset(x = seuratObj, idents = c("Control", "Hpoly.Day10"))
Idents(testObj) = 'clusterRes'
miloRes = getMilo(testObj)$da_results %>% 
  filter(ident_fraction > 0.8) %>%
    mutate(direct = ifelse( logFC > 0  & SpatialFDR < 0.1, "milo_more", "neutral"),
           direct = ifelse( logFC < 0  & SpatialFDR < 0.1, "milo_less", direct)) %>% 
    group_by(ident, direct) %>% 
    summarise(n = n()) %>% 
    pivot_wider(names_from = direct, values_from = n) %>% 
    replace(is.na(.), 0)%>% 
    dplyr::rename(cluster = ident)
  if ("milo_less" %!in% colnames(miloRes)) {miloRes$milo_less = rep(0, nrow(miloRes))}
  if ("milo_more" %!in% colnames(miloRes)) {miloRes$milo_more = rep(0, nrow(miloRes))}
```

```{r}
## results
all_resC10 = data.frame(cluster = rownames(betabin_null$LRT_pvals), truth = c("N","P","N","P","N","N","P","P")) %>% 
  mutate(betabin_null_pvals = as.vector(betabin_null$LRT_pvals),
         wtoPhi_emSVM_pvals = as.vector(wtoPhi_emSVM$LRT_pvals),
         estPhi_null_pvals = as.vector(estPhi_null$LRT_pvals),
         estPhi_emSVM_pvals = as.vector(estPhi_emSVM$LRT_pvals),
         fisher_pvals = fisher_pvals,
         scDC_pvals = scDCRes$p.value) %>%
  merge(speckleRes, by = "cluster") %>% 
  mutate(betabin_null_Res = ifelse(betabin_null_pvals < 0.01, "***", "n.s."),
         betabin_null_Res = ifelse(0.01<betabin_null_pvals&betabin_null_pvals<0.05, "**", betabin_null_Res),
         betabin_null_Res = ifelse(0.05<betabin_null_pvals&betabin_null_pvals<0.1, "*", betabin_null_Res)) %>% 
  #dplyr::select(truth, betabin_null_Res, betabin_null_pvals)
  mutate(wtoPhi_emSVM_Res = ifelse(wtoPhi_emSVM_pvals < 0.01, "***", "n.s."),
         wtoPhi_emSVM_Res = ifelse(0.01<wtoPhi_emSVM_pvals&wtoPhi_emSVM_pvals<0.05, "**", wtoPhi_emSVM_Res),
         wtoPhi_emSVM_Res = ifelse(0.05<wtoPhi_emSVM_pvals&wtoPhi_emSVM_pvals<0.1, "*", wtoPhi_emSVM_Res)) %>%
  #dplyr::select(truth, wtoPhi_emSVM_Res, wtoPhi_emSVM_pvals)
  mutate(estPhi_null_Res = ifelse(estPhi_null_pvals < 0.01, "***", "n.s."),
         estPhi_null_Res = ifelse(0.01<estPhi_null_pvals&estPhi_null_pvals<0.05, "**", estPhi_null_Res),
         estPhi_null_Res = ifelse(0.05<estPhi_null_pvals&estPhi_null_pvals<0.1, "*", estPhi_null_Res)) %>%
  mutate(estPhi_emSVM_Res = ifelse(estPhi_emSVM_pvals < 0.01, "***", "n.s."),
         estPhi_emSVM_Res = ifelse(0.01<estPhi_emSVM_pvals&estPhi_emSVM_pvals<0.05, "**", estPhi_emSVM_Res),
         estPhi_emSVM_Res = ifelse(0.05<estPhi_emSVM_pvals&estPhi_emSVM_pvals<0.1, "*", estPhi_emSVM_Res)) %>%
  mutate(fisher_Res = ifelse(fisher_pvals < 0.01, "***", "n.s."),
         fisher_Res = ifelse(0.01<fisher_pvals&fisher_pvals<0.05, "**", fisher_Res),
         fisher_Res = ifelse(0.05<fisher_pvals&fisher_pvals<0.1, "*", fisher_Res)) %>%
  #dplyr::select(truth, fisher_Res, fisher_pvals)
  mutate(scDC_Res = ifelse(scDC_pvals < 0.01, "***", "n.s."),
         scDC_Res = ifelse(0.01<scDC_pvals&scDC_pvals<0.05, "**", scDC_Res),
         scDC_Res = ifelse(0.05<scDC_pvals&scDC_pvals<0.1, "*", scDC_Res)) %>%
  #dplyr::select(truth, scDC_Res, scDC_pvals)
  mutate(speckle_Res = ifelse(speckle_pvals < 0.01, "***", "n.s."),
         speckle_Res = ifelse(0.01<speckle_pvals&speckle_pvals<0.05, "**", speckle_Res),
         speckle_Res = ifelse(0.05<speckle_pvals&speckle_pvals<0.1, "*", speckle_Res)) %>% 
  merge(miloRes, by = "cluster")


all_resC10 %>% 
  dplyr::select(cluster, truth, contains("_Res"))
```

control vs Salmonella

```{r}
## data convert
numb_cond1 = exp7_data %>% 
  filter(condition == "Control") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
numb_cond2 = exp7_data %>% 
  filter(condition == "Salmonella") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
exp7_CS = exp7_data %>% 
  filter(condition == "Control" | condition == "Salmonella") %>% mutate(clusterRes = as.factor(clusterRes))
```

Test part

```{r}
data("Haber2017")
countM = rbind(Haber2017$count_ctrl, Haber2017$count_Salma) %>% as.matrix()
designM = data.frame(condition = c(rep("g1", nrow(Haber2017$count_ctrl)), rep("g2", nrow(Haber2017$count_Salma))))
#betabin_null = dcats_GLM(countM, designM)
#wtoPhi_emSVM = dcats_GLM(countM, designM, similarity_mat = Haber2017$svm_mat, n_samples = NULL)
betabin_null = eachPhi(countM, designM)
wtoPhi_emSVM = eachPhi(countM, designM, similarity_mat = Haber2017$svm_mat)
estPhi = getPhi(countM, designM)
estPhi_null = dcats_GLM(countM, designM, fix_phi = estPhi)
estPhi_emSVM = dcats_GLM(countM, designM, similarity_mat = Haber2017$svm_mat, fix_phi = estPhi)
## Fisher's exact test
fisher_pvals = getFisher(as.matrix(Haber2017$count_ctrl), as.matrix(Haber2017$count_Salma))
## speckle
speckleRes = propeller(clusters = exp7_CS$clusterRes, sample = exp7_CS$batch, group = exp7_CS$condition) %>% 
  dplyr::rename(cluster = BaselineProp.clusters,
                      speckle_pvals = FDR) %>%
  dplyr::select(cluster, speckle_pvals)
## scDC
cluster_num = dim(numb_cond1)[2]
res_scDC <- scDC_noClustering(cellTypes = exp7_CS$clusterRes, exp7_CS$batch, calCI = TRUE, calCI_method = c("percentile", "BCa", "multinom"),nboot = 1000, verbose = FALSE)
res_GLM <- fitGLM(res_scDC, c(rep("cond1",dim(numb_cond1)[1]*cluster_num),rep("cond2",dim(numb_cond2)[1]*cluster_num)), pairwise = FALSE, fixed_only = TRUE, verbose = FALSE)
scDCRes_temp = summary(res_GLM$pool_res_fixed)
scDCRes = scDCRes_temp[c(cluster_num+1,(dim(scDCRes_temp)[1]-cluster_num+2):dim(scDCRes_temp)[1]),]
```

```{r}
plot(colSums(Haber2017$count_ctrl)+colSums(Haber2017$count_Salma), betabin_null$fm0_phi)
abline(h = phi, col = 'red')
plot(colSums(Haber2017$count_ctrl)+colSums(Haber2017$count_Salma), betabin_null$fm1_phi)
abline(h = phi, col = 'red')
plot(colSums(Haber2017$count_ctrl)+colSums(Haber2017$count_Salma), wtoPhi_emSVM$fm0_phi)
abline(h = phi, col = 'red')
plot(colSums(Haber2017$count_ctrl)+colSums(Haber2017$count_Salma), wtoPhi_emSVM$fm1_phi)
abline(h = phi, col = 'red')
plot(colSums(Haber2017$count_ctrl)+colSums(Haber2017$count_Salma), adjPhi)
abline(h = phi, col = 'red')
```

```{r}
set.seed(123)
testObj = subset(x = seuratObj, idents = c("Control", "Salmonella"))
Idents(testObj) = 'clusterRes'
miloRes = getMilo(testObj)$da_results %>% 
  filter(ident_fraction > 0.8) %>%
    mutate(direct = ifelse( logFC > 0  & SpatialFDR < 0.1, "milo_more", "neutral"),
           direct = ifelse( logFC < 0  & SpatialFDR < 0.1, "milo_less", direct)) %>% 
    group_by(ident, direct) %>% 
    summarise(n = n()) %>% 
    pivot_wider(names_from = direct, values_from = n) %>% 
    replace(is.na(.), 0)%>% 
    dplyr::rename(cluster = ident)
  if ("milo_less" %!in% colnames(miloRes)) {miloRes$milo_less = rep(0, nrow(miloRes))}
  if ("milo_more" %!in% colnames(miloRes)) {miloRes$milo_more = rep(0, nrow(miloRes))}
```

```{r}
## results
all_resCS = data.frame(cluster = rownames(betabin_null$LRT_pvals), truth = c("N","P","N","N","P","P","P","N")) %>% 
   mutate(betabin_null_pvals = as.vector(betabin_null$LRT_pvals),
         wtoPhi_emSVM_pvals = as.vector(wtoPhi_emSVM$LRT_pvals),
         estPhi_null_pvals = as.vector(estPhi_null$LRT_pvals),
         estPhi_emSVM_pvals = as.vector(estPhi_emSVM$LRT_pvals),
         fisher_pvals = fisher_pvals,
         scDC_pvals = scDCRes$p.value) %>%
  merge(speckleRes, by = "cluster") %>% 
  mutate(betabin_null_Res = ifelse(betabin_null_pvals < 0.01, "***", "n.s."),
         betabin_null_Res = ifelse(0.01<betabin_null_pvals&betabin_null_pvals<0.05, "**", betabin_null_Res),
         betabin_null_Res = ifelse(0.05<betabin_null_pvals&betabin_null_pvals<0.1, "*", betabin_null_Res)) %>% 
  #dplyr::select(truth, betabin_null_Res, betabin_null_pvals)
  mutate(wtoPhi_emSVM_Res = ifelse(wtoPhi_emSVM_pvals < 0.01, "***", "n.s."),
         wtoPhi_emSVM_Res = ifelse(0.01<wtoPhi_emSVM_pvals&wtoPhi_emSVM_pvals<0.05, "**", wtoPhi_emSVM_Res),
         wtoPhi_emSVM_Res = ifelse(0.05<wtoPhi_emSVM_pvals&wtoPhi_emSVM_pvals<0.1, "*", wtoPhi_emSVM_Res)) %>%
  #dplyr::select(truth, wtoPhi_emSVM_Res, wtoPhi_emSVM_pvals)
  mutate(estPhi_null_Res = ifelse(estPhi_null_pvals < 0.01, "***", "n.s."),
         estPhi_null_Res = ifelse(0.01<estPhi_null_pvals&estPhi_null_pvals<0.05, "**", estPhi_null_Res),
         estPhi_null_Res = ifelse(0.05<estPhi_null_pvals&estPhi_null_pvals<0.1, "*", estPhi_null_Res)) %>%
  mutate(estPhi_emSVM_Res = ifelse(estPhi_emSVM_pvals < 0.01, "***", "n.s."),
         estPhi_emSVM_Res = ifelse(0.01<estPhi_emSVM_pvals&estPhi_emSVM_pvals<0.05, "**", estPhi_emSVM_Res),
         estPhi_emSVM_Res = ifelse(0.05<estPhi_emSVM_pvals&estPhi_emSVM_pvals<0.1, "*", estPhi_emSVM_Res)) %>%
  mutate(fisher_Res = ifelse(fisher_pvals < 0.01, "***", "n.s."),
         fisher_Res = ifelse(0.01<fisher_pvals&fisher_pvals<0.05, "**", fisher_Res),
         fisher_Res = ifelse(0.05<fisher_pvals&fisher_pvals<0.1, "*", fisher_Res)) %>%
  #dplyr::select(truth, fisher_Res, fisher_pvals)
  mutate(scDC_Res = ifelse(scDC_pvals < 0.01, "***", "n.s."),
         scDC_Res = ifelse(0.01<scDC_pvals&scDC_pvals<0.05, "**", scDC_Res),
         scDC_Res = ifelse(0.05<scDC_pvals&scDC_pvals<0.1, "*", scDC_Res)) %>%
  #dplyr::select(truth, scDC_Res, scDC_pvals)
  mutate(speckle_Res = ifelse(speckle_pvals < 0.01, "***", "n.s."),
         speckle_Res = ifelse(0.01<speckle_pvals&speckle_pvals<0.05, "**", speckle_Res),
         speckle_Res = ifelse(0.05<speckle_pvals&speckle_pvals<0.1, "*", speckle_Res)) %>% 
  merge(miloRes, by = "cluster")


all_resCS %>% 
  dplyr::select(cluster, truth, contains("_Res"))
```

```{r}
save(all_resC3, all_resC10, all_resCS, file = "D:/Data/DCATS/real_world/Haber2017_res.RData")
```

## real-world data 5: Exp6-demuxlet

### store data

```{r, eval=FALSE}
## save data as demonstarate data
Kang2017 = list(numb_cond1 = as.matrix(numb_cond1), numb_cond2 = as.matrix(numb_cond2), svm_mat = svm_mat, svmDF = exp6DF, source = "Multiplexed droplet single-cell RNA-sequencing using natural genetic variation")
save(Kang2017, file = "./data/real_world/Kang2017.RData")
```

```{r, eval=FALSE}
graphs = seurat_intg@graphs$RNA_snn
labels = exp6DF$clusterRes
knn_mat = KNN_transition(graphs, labels)
order1 = colnames(numb_cond1)
simil_matK = knn_mat[order1, order1]
```


```{r}
exp6_clusterInfo = read.delim("./data/real_world/Exp6_Kang/GSE96583_batch2.total.tsne.df.tsv") %>% 
  dplyr::rename(clusterRes = cell, condition = stim) %>% 
  rownames_to_column("cell") %>% 
  filter(!is.na(clusterRes))
head(exp6_clusterInfo)
exp6_clusterInfo %>% group_by(condition, clusterRes) %>% 
  summarise(n = n())
## check whether exist duplicate barcode -> no
exp6_clusterInfo %>% 
  group_by(cell) %>% 
  summarise(n = n()) %>% 
  filter(n > 1)
```

```{r}
matrix1 = readMM(gzfile("./data/real_world/Exp6_Kang/GSM2560248_2.1.mtx.gz")) %>% 
  as.matrix()
matrix2 = readMM(gzfile("./data/real_world/Exp6_Kang/GSM2560249_2.2.mtx.gz")) %>% 
  as.matrix()
```

```{r}
data_dir1 = "./data/real_world/Exp6_Kang/GSM2560248"
list.files(data_dir1) # Should show barcodes.tsv, genes.tsv, and matrix.mtx
expression_matrix1 <- Read10X(data.dir = data_dir1)
seuratObj1 = CreateSeuratObject(counts = expression_matrix1)
seuratObj1 = AddMetaData(object = seuratObj1, metadata = rep("ctrl", dim(seuratObj1)[2]), col.name = 'sample')

data_dir2 = "./data/real_world/Exp6_Kang/GSM2560249"
list.files(data_dir2) # Should show barcodes.tsv, genes.tsv, and matrix.mtx
expression_matrix2 <- Read10X(data.dir = data_dir2)
seuratObj2 = CreateSeuratObject(counts = expression_matrix2)
seuratObj2 = AddMetaData(object = seuratObj2, metadata = rep("stim", dim(seuratObj2)[2]), col.name = 'sample')
```

```{r}
seuratObjL = c(sub1 = seuratObj1, sub2 = seuratObj2)
```

```{r}
anchors <- FindIntegrationAnchors(object.list = seuratObjL, dims = 1:30)
seurat_intg <- IntegrateData(anchorset = anchors, dims = 1:30)
seurat_intg <- ScaleData(seurat_intg)
seurat_intg <- RunPCA(seurat_intg, npcs = 30, verbose = FALSE)
```

```{r}
ctrl_repInfo <- read.delim("./data/real_world/Exp6_Kang/ye1.ctrl.8.10.sm.best")
stim_repInfo <- read.delim("./data/real_world/Exp6_Kang/ye2.stim.8.10.sm.best")
head(ctrl_repInfo)
head(stim_repInfo)

exp6_repInfo = rbind(ctrl_repInfo, stim_repInfo) %>% 
  dplyr::rename(cell = BARCODE, batch = BEST) %>% 
  filter(str_detect(batch, 'SNG')) %>% 
  dplyr::select(cell, batch)
head(exp6_repInfo)
```

```{r}
exp6DF_mul = seurat_intg@reductions$pca@cell.embeddings %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("cells") %>% 
  mutate(cell = str_replace(cells, "_[12]", "")) %>% 
  merge(exp6_clusterInfo) %>% 
  merge(exp6_repInfo) %>% 
  mutate(batch = str_c(condition, batch)) %>% 
  dplyr::select(-cell, -cluster, -multiplets, -tsne1, -tsne2, -ind)

testObj = subset(x = seurat_intg, cells = exp6DF_mul$cells)
testObj[['condition']] = exp6DF_mul$condition
testObj[['batch']] = exp6DF_mul$batch
testObj[['clusterRes']] = exp6DF_mul$clusterRes
Idents(testObj) = 'clusterRes'
exp6DF_mul = exp6DF_mul %>% dplyr::select(-cells)
head(exp6DF_mul)
```

```{r, eval=FALSE}
## data convert
numb_cond1 = exp6DF_mul %>% 
  filter(condition == "ctrl") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
numb_cond2 = exp6DF_mul %>% 
  filter(condition == "stim") %>% 
  group_by(clusterRes, batch) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = "clusterRes", values_from = "n") %>% 
  column_to_rownames(var = "batch")
```

Test part

```{r}
data("Kang2017")
countM = rbind(Kang2017$count_ctrl, Kang2017$count_stim) %>% as.matrix()
designM = data.frame(condition = c(rep("g1", nrow(Kang2017$count_ctrl)), rep("g2", nrow(Kang2017$count_stim))))
betabin_null = eachPhi(countM, designM)
wtoPhi_emSVM = eachPhi(countM, designM, similarity_mat = Haber2017$svm_mat)
estPhi = getPhi(countM, designM)
estPhi_null = dcats_GLM(countM, designM, fix_phi = estPhi)
estPhi_emSVM = dcats_GLM(countM, designM, similarity_mat = Haber2017$svm_mat, fix_phi = estPhi)
## Fisher's exact test
fisher_pvals = getFisher(as.matrix(Kang2017$count_ctrl), as.matrix(Kang2017$count_stim))
## speckle
speckleRes = speckle::propeller(clusters = exp6DF_mul$clusterRes, sample = exp6DF_mul$batch, group = exp6DF_mul$condition) %>% 
  dplyr::rename(cluster = BaselineProp.clusters,
                      speckle_pvals = FDR) %>%
  dplyr::select(cluster, speckle_pvals)
## scDC
cluster_num = dim(Kang2017$count_ctrl)[2]
res_scDC <- scDC_noClustering(cellTypes = exp6DF_mul$clusterRes, exp6DF_mul$batch, calCI = TRUE, calCI_method = c("percentile", "BCa", "multinom"),nboot = 1000, verbose = FALSE)
res_GLM <- fitGLM(res_scDC, c(rep("cond1",dim(Kang2017$count_ctrl)[1]*cluster_num),rep("cond2",dim(Kang2017$count_stim)[1]*cluster_num)), pairwise = FALSE, fixed_only = TRUE, verbose = FALSE)
scDCRes_temp = summary(res_GLM$pool_res_fixed)
scDCRes = scDCRes_temp[c(cluster_num+1,(dim(scDCRes_temp)[1]-cluster_num+2):dim(scDCRes_temp)[1]),]
```

```{r}
miloRes = getMilo(testObj)$da_results %>% 
  filter(ident_fraction > 0.8) %>%
    mutate(direct = ifelse( logFC > 0  & SpatialFDR < 0.1, "milo_more", "neutral"),
           direct = ifelse( logFC < 0  & SpatialFDR < 0.1, "milo_less", direct)) %>% 
    group_by(ident, direct) %>% 
    summarise(n = n()) %>% 
    pivot_wider(names_from = direct, values_from = n) %>% 
    replace(is.na(.), 0) %>% 
    dplyr::rename(cluster = ident)
  if ("milo_less" %!in% colnames(miloRes)) {miloRes$milo_less = rep(0, nrow(miloRes))}
  if ("milo_more" %!in% colnames(miloRes)) {miloRes$milo_more = rep(0, nrow(miloRes))}
```


```{r}
## results
all_res = data.frame(cluster = rownames(betabin_null$LRT_pvals), truth = rep("N", 8)) %>% 
  mutate(betabin_null_pvals = as.vector(betabin_null$LRT_pvals),
         wtoPhi_emSVM_pvals = as.vector(wtoPhi_emSVM$LRT_pvals),
         estPhi_null_pvals = as.vector(estPhi_null$LRT_pvals),
         estPhi_emSVM_pvals = as.vector(estPhi_emSVM$LRT_pvals),
         fisher_pvals = fisher_pvals,
         scDC_pvals = scDCRes$p.value) %>%
  merge(speckleRes, by = "cluster") %>% 
  mutate(betabin_null_Res = ifelse(betabin_null_pvals < 0.01, "***", "n.s."),
         betabin_null_Res = ifelse(0.01<betabin_null_pvals&betabin_null_pvals<0.05, "**", betabin_null_Res),
         betabin_null_Res = ifelse(0.05<betabin_null_pvals&betabin_null_pvals<0.1, "*", betabin_null_Res)) %>% 
  #dplyr::select(truth, betabin_null_Res, betabin_null_pvals)
  mutate(wtoPhi_emSVM_Res = ifelse(wtoPhi_emSVM_pvals < 0.01, "***", "n.s."),
         wtoPhi_emSVM_Res = ifelse(0.01<wtoPhi_emSVM_pvals&wtoPhi_emSVM_pvals<0.05, "**", wtoPhi_emSVM_Res),
         wtoPhi_emSVM_Res = ifelse(0.05<wtoPhi_emSVM_pvals&wtoPhi_emSVM_pvals<0.1, "*", wtoPhi_emSVM_Res)) %>%
  #dplyr::select(truth, wtoPhi_emSVM_Res, wtoPhi_emSVM_pvals)
  mutate(estPhi_null_Res = ifelse(estPhi_null_pvals < 0.01, "***", "n.s."),
         estPhi_null_Res = ifelse(0.01<estPhi_null_pvals&estPhi_null_pvals<0.05, "**", estPhi_null_Res),
         estPhi_null_Res = ifelse(0.05<estPhi_null_pvals&estPhi_null_pvals<0.1, "*", estPhi_null_Res)) %>%
  mutate(estPhi_emSVM_Res = ifelse(estPhi_emSVM_pvals < 0.01, "***", "n.s."),
         estPhi_emSVM_Res = ifelse(0.01<estPhi_emSVM_pvals&estPhi_emSVM_pvals<0.05, "**", estPhi_emSVM_Res),
         estPhi_emSVM_Res = ifelse(0.05<estPhi_emSVM_pvals&estPhi_emSVM_pvals<0.1, "*", estPhi_emSVM_Res)) %>%
  mutate(fisher_Res = ifelse(fisher_pvals < 0.01, "***", "n.s."),
         fisher_Res = ifelse(0.01<fisher_pvals&fisher_pvals<0.05, "**", fisher_Res),
         fisher_Res = ifelse(0.05<fisher_pvals&fisher_pvals<0.1, "*", fisher_Res)) %>%
  #dplyr::select(truth, fisher_Res, fisher_pvals)
  mutate(scDC_Res = ifelse(scDC_pvals < 0.01, "***", "n.s."),
         scDC_Res = ifelse(0.01<scDC_pvals&scDC_pvals<0.05, "**", scDC_Res),
         scDC_Res = ifelse(0.05<scDC_pvals&scDC_pvals<0.1, "*", scDC_Res)) %>%
  #dplyr::select(truth, scDC_Res, scDC_pvals)
  mutate(speckle_Res = ifelse(speckle_pvals < 0.01, "***", "n.s."),
         speckle_Res = ifelse(0.01<speckle_pvals&speckle_pvals<0.05, "**", speckle_Res),
         speckle_Res = ifelse(0.05<speckle_pvals&speckle_pvals<0.1, "*", speckle_Res)) %>% 
  merge(miloRes, by = "cluster")


all_res %>% 
  dplyr::select(cluster, truth, contains("_Res"))
save(all_res, file = "D:/Data/DCATS/real_world/Kang2017_res.RData")
```

## Ren 2021

COVID-19 immune features revealed by a large-scale single-cell transcriptome atlas

First look at PBMC data.

### save data

```{r, eval=FALSE}
Ren2021 = list(countM = countM, designM = designM, source = "COVID-19 immune features revealed by a large-scale single-cell transcriptome atlas")
save(Ren2021, file = "D:/Data/DCATS/Ren2021.RData")
```


```{r}
library(readxl)
Ren2021_cell = read.csv(gzfile("D:/Data/DCATS/real_world/Ren_2021/cell_annotation.csv.gz")) %>% 
  janitor::clean_names()
Ren2021_donors = read_excel("D:/Data/DCATS/real_world/Ren_2021/donors_info.xlsx") %>% 
  janitor::clean_names()
Ren2021_pca = read.csv("D:/Data/DCATS/real_world/Ren_2021/pca_subsample.csv") %>% 
  dplyr::select(-X, -sampleID) %>% 
  rename(cell_name = barcode)
```

```{r}
pbmc_sampleL = Ren2021_donors %>% 
  filter(sample_type == "frozen PBMC" | sample_type == "fresh PBMC") %>% 
  .$sample_id %>% 
  unique()

sample_patient = Ren2021_donors %>% dplyr::select(sample_id, patient_id)
```

### major type

```{r}
countM = Ren2021_cell %>% 
  filter(sample_id %in% pbmc_sampleL) %>% 
  group_by(sample_id, major_type) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = major_type, values_from = n) %>% 
  arrange(sample_id) %>% 
  filter(sample_id != 'S-M032') %>% 
  column_to_rownames("sample_id") %>% 
  replace(is.na(.), 0) %>% 
  as.matrix()
```

```{r}
designM = Ren2021_donors %>% 
  filter(sample_id %in% pbmc_sampleL) %>% 
  mutate(state = ifelse(sample_time == "control", "control", str_c(co_vid_19_severity, "_", sample_time)),
         age = as.numeric(age)) %>% 
  dplyr::select(sample_id, age, sex, sample_type, state) %>% 
  #unique() %>% 
  arrange(sample_id) %>% 
  filter(sample_id != 'S-M032') %>% 
  column_to_rownames("sample_id")
designM
```

```{r}
simil_mat = create_simMat(dim(countM)[2], confuse_rate=0.1)
betabin_null = dcats_GLM(countM, designM, base_model = 'FULL')
wtoPhi_emSVM = dcats_GLM(countM, designM, similarity_mat = simil_mat, n_samples = NULL, base_model = 'FULL')
phi = getPhi(countM, designM)
estPhi_null = dcats_GLM(countM, designM, fix_phi = phi, base_model = 'FULL')
estPhi_emSVM = dcats_GLM(countM, designM, similarity_mat = simil_mat, n_samples = NULL, fix_phi = phi, base_model = 'FULL')

#ifelse(res_GLM$LRT_pvals < 0.05, "P", "N") 
ifelse(res_GLM$pvals < 0.05, "P", "N") 
#print(res_GLM$LR_vals)
print(res_GLM$pvals)
```

Separate into different group comparison

control vs mild/moderate_progression

```{r}
designM_sub = designM %>% 
  filter(state == "control" | state == "mild/moderate_progression")
countM_sub = countM[rownames(designM_sub),]

simil_mat = create_simMat(dim(countM_sub)[2], confuse_rate=0.1)
#betabin_null = dcats_GLM(countM_sub, designM_sub, base_model = 'FULL')
#wtoPhi_emSVM = dcats_GLM(countM_sub, designM_sub, similarity_mat = simil_mat, n_samples = NULL, base_model = 'FULL')
betabin_null = eachPhi(countM_sub, designM_sub, base_model = 'FULL')
wtoPhi_emSVM = eachPhi(countM_sub, designM_sub, similarity_mat = simil_mat, n_samples = NULL, base_model = 'FULL')
phi = getPhi(countM_sub, designM_sub)
estPhi_null = dcats_GLM(countM_sub, designM_sub, fix_phi = phi, base_model = 'FULL')
estPhi_emSVM = dcats_GLM(countM_sub, designM_sub, similarity_mat = simil_mat, n_samples = NULL, fix_phi = phi, base_model = 'FULL')
betabin_null$LRT_pvals < 0.05
wtoPhi_emSVM$LRT_pvals < 0.05
estPhi_null$LRT_pvals < 0.05
estPhi_emSVM$LRT_pvals < 0.05
```

```{r}
plot(colSums(countM_sub), betabin_null$fm0_phi[,1])
abline(h = phi, col = 'red')
plot(colSums(countM_sub), betabin_null$fm1_phi[,1])
abline(h = phi, col = 'red')
plot(colSums(countM_sub), wtoPhi_emSVM$fm0_phi[,1])
abline(h = phi, col = 'red')
plot(colSums(countM_sub), wtoPhi_emSVM$fm1_phi[,1])
abline(h = phi, col = 'red')
```


Separate into different group comparison

control vs severe/critical_progression

```{r}
designM_sub = designM %>% 
  filter(state == "control" | state == "severe/critical_progression")
countM_sub = countM[rownames(designM_sub),]

simil_mat = create_simMat(dim(countM_sub)[2], confuse_rate=0.1)
#betabin_null = dcats_GLM(countM_sub, designM_sub, base_model = 'FULL')
#wtoPhi_emSVM = dcats_GLM(countM_sub, designM_sub, similarity_mat = simil_mat, n_samples = NULL, base_model = 'FULL')
betabin_null = eachPhi(countM_sub, designM_sub, base_model = 'FULL')
wtoPhi_emSVM = eachPhi(countM_sub, designM_sub, similarity_mat = simil_mat, n_samples = NULL, base_model = 'FULL')
phi = getPhi(countM_sub, designM_sub)
estPhi_null = dcats_GLM(countM_sub, designM_sub, fix_phi = phi, base_model = 'FULL')
estPhi_emSVM = dcats_GLM(countM_sub, designM_sub, similarity_mat = simil_mat, n_samples = NULL, fix_phi = phi, base_model = 'FULL')
betabin_null$LRT_pvals < 0.05
wtoPhi_emSVM$LRT_pvals < 0.05
estPhi_null$LRT_pvals < 0.05
estPhi_emSVM$LRT_pvals < 0.05
```

```{r}
plot(colSums(countM_sub), betabin_null$fm0_phi[,1])
abline(h = phi, col = 'red')
plot(colSums(countM_sub), betabin_null$fm1_phi[,1])
abline(h = phi, col = 'red')
plot(colSums(countM_sub), wtoPhi_emSVM$fm0_phi[,1])
abline(h = phi, col = 'red')
plot(colSums(countM_sub), wtoPhi_emSVM$fm1_phi[,1])
abline(h = phi, col = 'red')
```

```{r}
count_mat_sub %>% as.data.frame() %>% 
  cbind(design_mat_sub) %>% 
  rownames_to_column("sample_id") %>% 
  pivot_longer(names_to = "cell_type", values_to = "n", B:Neu) %>% 
  group_by(sample_id) %>% 
  mutate(prop = n/sum(n)) %>% 
  ggplot(aes(x = cell_type, y = prop)) +
    geom_boxplot(aes(color = state))
```

mild/moderate_convalescence vs mild/moderate_progression

```{r}
designM_sub = designM %>% 
  filter(state == "mild/moderate_convalescence" | state == "mild/moderate_progression")
countM_sub = countM[rownames(designM_sub),]

simil_mat = create_simMat(dim(countM_sub)[2], confuse_rate=0.1)
#betabin_null = dcats_GLM(countM_sub, designM_sub, base_model = 'FULL')
#wtoPhi_emSVM = dcats_GLM(countM_sub, designM_sub, similarity_mat = simil_mat, n_samples = NULL, base_model = 'FULL')
betabin_null = eachPhi(countM_sub, designM_sub, base_model = 'FULL')
wtoPhi_emSVM = eachPhi(countM_sub, designM_sub, similarity_mat = simil_mat, n_samples = NULL, base_model = 'FULL')
phi = getPhi(countM_sub, designM_sub)
estPhi_null = dcats_GLM(countM_sub, designM_sub, fix_phi = phi, base_model = 'FULL')
estPhi_emSVM = dcats_GLM(countM_sub, designM_sub, similarity_mat = simil_mat, n_samples = NULL, fix_phi = phi, base_model = 'FULL')
betabin_null$LRT_pvals < 0.05
wtoPhi_emSVM$LRT_pvals < 0.05
estPhi_null$LRT_pvals < 0.05
estPhi_emSVM$LRT_pvals < 0.05
```

```{r}
colSums(countM_sub)[order(colSums(countM_sub))]
plot(colSums(countM_sub), betabin_null$fm0_phi[,1])
abline(h = phi, col = 'red')
plot(colSums(countM_sub), betabin_null$fm1_phi[,1])
abline(h = phi, col = 'red')
plot(colSums(countM_sub), wtoPhi_emSVM$fm0_phi[,1])
abline(h = phi, col = 'red')
plot(colSums(countM_sub), wtoPhi_emSVM$fm1_phi[,1])
abline(h = phi, col = 'red')
```

```{r}
count_mat_sub %>% as.data.frame() %>% 
  cbind(design_mat_sub) %>% 
  rownames_to_column("sample_id") %>% 
  pivot_longer(names_to = "cell_type", values_to = "n", B:Neu) %>% 
  group_by(sample_id) %>% 
  mutate(prop = n/sum(n)) %>% 
  ggplot(aes(x = cell_type, y = prop)) +
    geom_boxplot(aes(color = state))
```

severe/critical_convalescence vs severe/critical_progression

```{r}
designM_sub = designM %>% 
  filter(state == "severe/critical_convalescence" | state == "severe/critical_progression")
countM_sub = countM[rownames(designM_sub),]

simil_mat = create_simMat(dim(countM_sub)[2], confuse_rate=0.1)
#betabin_null = dcats_GLM(countM_sub, designM_sub, base_model = 'FULL')
#wtoPhi_emSVM = dcats_GLM(countM_sub, designM_sub, similarity_mat = simil_mat, n_samples = NULL, base_model = 'FULL')
betabin_null = eachPhi(countM_sub, designM_sub, base_model = 'FULL')
wtoPhi_emSVM = eachPhi(countM_sub, designM_sub, similarity_mat = simil_mat, n_samples = NULL, base_model = 'FULL')
phi = getPhi(countM_sub, designM_sub)
estPhi_null = dcats_GLM(countM_sub, designM_sub, fix_phi = phi, base_model = 'FULL')
estPhi_emSVM = dcats_GLM(countM_sub, designM_sub, similarity_mat = simil_mat, n_samples = NULL, fix_phi = phi, base_model = 'FULL')
betabin_null$LRT_pvals < 0.05
wtoPhi_emSVM$LRT_pvals < 0.05
estPhi_null$LRT_pvals < 0.05
estPhi_emSVM$LRT_pvals < 0.05
```

```{r}
colSums(countM_sub)[order(colSums(countM_sub))]
plot(colSums(countM_sub), betabin_null$fm0_phi[,1])
abline(h = phi, col = 'red')
plot(colSums(countM_sub), betabin_null$fm1_phi[,1])
abline(h = phi, col = 'red')
plot(colSums(countM_sub), wtoPhi_emSVM$fm0_phi[,1])
abline(h = phi, col = 'red')
plot(colSums(countM_sub), wtoPhi_emSVM$fm1_phi[,1])
abline(h = phi, col = 'red')
```

```{r}
count_mat_sub %>% as.data.frame() %>% 
  cbind(design_mat_sub) %>% 
  rownames_to_column("sample_id") %>% 
  pivot_longer(names_to = "cell_type", values_to = "n", B:Neu) %>% 
  group_by(sample_id) %>% 
  mutate(prop = n/sum(n)) %>% 
  ggplot(aes(x = cell_type, y = prop)) +
    geom_boxplot(aes(color = state))
```

mild/moderate_progression vs severe/critical_progression

```{r}
design_mat_sub = design_mat %>% 
  filter(state == "mild/moderate_progression" | state == "severe/critical_progression")
count_mat_sub = count_mat[rownames(design_mat_sub),]

simil_mat = create_simMat(dim(count_mat_sub)[2], confuse_rate=0.1)
res_GLM = dcats_GLM(count_mat = count_mat_sub, design_mat = design_mat_sub, similarity_mat = simil_mat)
#ifelse(res_GLM$LRT_pvals < 0.05, "P", "N") 
ifelse(res_GLM$pvals < 0.05, "P", "N") 
#print(res_GLM$LR_vals)
print(res_GLM$pvals)
```

```{r}
count_mat_sub %>% as.data.frame() %>% 
  cbind(design_mat_sub) %>% 
  rownames_to_column("sample_id") %>% 
  pivot_longer(names_to = "cell_type", values_to = "n", B:Neu) %>% 
  group_by(sample_id) %>% 
  mutate(prop = n/sum(n)) %>% 
  ggplot(aes(x = cell_type, y = prop)) +
    geom_boxplot(aes(color = state))
```



## sub type

First, check the number of sub-celltype. Some sub-cell type contains less cell count. Those cell types including DC_c1-CLEC9A, NK_c02-NCAM1, NK_c03-MKI67, T_CD4_c13-MKI67-CCL5_l, T_CD4_c14-MKI67-CCL5_h, T_CD8_c12-MKI67-TYROBP, T_CD8_c13-HAVCR2, Macro_c1-C1QC, B_c05-MZB1-XBP1, B_c06-MKI67, T_CD4_c10-IFNG, T_CD8_c11-MKI67-FOS, all Neu, Macro

```{r}
Ren2021_cell %>% 
  filter(sample_id %in% pbmc_sampleL) %>% 
  group_by(sample_id, celltype) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = celltype, values_from = n)
```

```{r}
library(tidymodels)
svm_input = Ren2021_cell %>% 
  filter(cell_name %in% Ren2021_pca$cell_name) %>% 
  plyr::join(Ren2021_pca) %>% 
  mutate(clusterRes = ifelse(major_type %in% c("Neu", "Macro"), major_type, celltype),
         clusterRes = ifelse(celltype %in% c("DC_c1-CLEC9A", "NK_c02-NCAM1", "NK_c03-MKI67", "T_CD4_c13-MKI67-CCL5_l", "T_CD4_c14-MKI67-CCL5_h", "T_CD8_c12-MKI67-TYROBP", "T_CD8_c13-HAVCR2", "B_c05-MZB1-XBP1", "B_c06-MKI67", "T_CD4_c10-IFNG", "T_CD8_c11-MKI67-FOS"), "others", clusterRes)) %>% 
  select(-cell_name, -sample_id, -major_type, -celltype)
svm_mat = svm_simMat(svm_input)
```


```{r}
Ren2021_cell %>% 
  mutate(clusterRes = ifelse(major_type %in% c("Neu", "Macro"), major_type, celltype)) %>% 
  select(-cell_namem, -sample_id, -major_type, -celltype)
```


```{r}
Neu_count = Ren2021_cell %>% 
  merge(sample_patient) %>% 
  filter(sample_id %in% pbmc_sampleL) %>% 
  group_by(patient_id, celltype) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = celltype, values_from = n) %>% 
  column_to_rownames("patient_id") %>% 
  select(starts_with("Neu")) %>% 
  replace(is.na(.), 0) %>% 
  rowSums()
Macro_count = Ren2021_cell %>% 
  merge(sample_patient) %>% 
  filter(sample_id %in% pbmc_sampleL) %>% 
  group_by(patient_id, celltype) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = celltype, values_from = n) %>% 
  column_to_rownames("patient_id") %>% 
  select(starts_with("Macro")) %>% 
  replace(is.na(.), 0) %>% 
  rowSums()
```

```{r}
count_mat = Ren2021_cell %>% 
  merge(sample_patient) %>% 
  filter(sample_id %in% pbmc_sampleL) %>% 
  group_by(patient_id, celltype) %>% 
  summarise(n = n()) %>% 
  pivot_wider(names_from = celltype, values_from = n) %>% 
  column_to_rownames("patient_id") %>% 
  select(-starts_with("Neu"), -starts_with("Macro")) %>% 
  mutate(Neu = Neu_count,
         Macro = Macro_count)
```

## Schiller 2019 data

https://www.nature.com/articles/s41467-019-08831-9#data-availability

An atlas of the aging lung mapped by single cell transcriptomics and deep tissue proteomics

Sample size: whole lungs samples from 3-month-old mice (n = 8) and 24-month-old mice (m = 7). In total 14813 cells (7672 young, 7141 old).

Key findings in this paper:

1) Additionally, we noticed one cluster of mainly proliferating cells showing high expression levels for S and G2M cell-cycle marker genes (Supplementary Fig. 3a and b). Young mice showed a higher fraction of cells in this cluster compared to old mice

2) The ciliated cells were increased in old mice (Fig. 4h), leading to a significantly altered ratio of club to ciliated cells in aged mouse airways (Fig. 4i). validated this finding in situ immunostainings of Foxj1 (ciliated cell marker) and CC10 (club cell marker)

```{r}
load("/storage/holab/linxy/DCATS/real_world/Schiller_2019/GSE124872_raw_counts_single_cell.RData")
meta_Schiller2019 = read_csv("/storage/holab/linxy/DCATS/real_world/Schiller_2019/GSE124872_Angelidis_2018_metadata.csv")
seurat_Schiller2019 = CreateSeuratObject(raw_counts)
```

```{r}
Cells(raw_counts) %>% head()
meta_Schiller2019$...1 %>% head()   ## barcodes are not the same
summary(meta_Schiller2019$identifier %>% as.factor())
cellDF = data.frame(cell = Cells(raw_counts)) %>% 
  mutate(cellID = cell) %>% 
  separate(cellID, c("sample", "barcode"), sep = ":")
summary(cellDF$sample %>% as.factor())
sampleDF = data.frame(sample = c("old_1", "old_2", "old_3", "old_4", "old_5", "old_6", "old_7", "young_1", "young_2", "young_3", "young_4", "young_5", "young_6", "young_7", "young_8"), identifier = c("muc3838", "muc3839", "muc3840", "muc3841", "muc4166", "muc4167", "muc4168", "muc4169", "muc4170", "muc4172", "muc4173", "muc4174", "muc4175", "muc4654", "muc4657"))
meta_Schiller2019 = meta_Schiller2019 %>% left_join(sampleDF) %>% 
  rename(cell = ...1) %>%
  mutate(barcode = str_remove(cell, "muc[0-9][0-9][0-9][0-9]:muc[0-9][0-9][0-9][0-9]:")) %>% 
  mutate(cell = str_c(sample, ":", barcode))
```

```{r}
meta_Schiller2019 = cellDF %>% 
  left_join(meta_Schiller2019)

seurat_Schiller2019@meta.data['cell_type'] = meta_Schiller2019$celltype
seurat_Schiller2019@meta.data['sample'] = meta_Schiller2019$sample
seurat_Schiller2019@meta.data['cluster'] = meta_Schiller2019$cluster
```

```{r}
seurat_Schiller2019 <- NormalizeData(seurat_Schiller2019)
seurat_Schiller2019 <- FindVariableFeatures(seurat_Schiller2019, selection.method = "vst", nfeatures = 2000)
seurat_Schiller2019 <- ScaleData(seurat_Schiller2019, verbose = FALSE)
seurat_Schiller2019 <- RunPCA(seurat_Schiller2019, npcs = 30, verbose = FALSE)
seurat_Schiller2019 <- RunUMAP(seurat_Schiller2019, reduction = "pca", dims = 1:30)
seurat_Schiller2019 <- FindNeighbors(seurat_Schiller2019, reduction = "pca", dims = 1:30)
seurat_Schiller2019 <- FindClusters(seurat_Schiller2019, resolution = 0.5)
```

```{r}
DimPlot(seurat_Schiller2019, group.by = 'sample') ## No batch effect, don't need to do integration
svmDF_Schiller2019 = seurat_Schiller2019@reductions$pca@cell.embeddings %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("cell") %>% 
  separate(cell, c("batch", "barcode"), sep = ":") %>% 
  mutate(condition = str_remove(batch, "_[0-9]")) %>% 
  dplyr::select(-barcode, -batch)

svmDF_Schiller2019['clusterRes'] = seurat_Schiller2019$cell_type 
svm_mat = svm_simMat(svmDF_Schiller2019)
```

```{r}
svm_simMat = function(dataframe){
  set.seed(123)
  dataframe$clusterRes = as.factor(dataframe$clusterRes)
  row_num = nrow(dataframe)
  pred = rep(NA, row_num)
  idx1 = sample(1:row_num, round(row_num/5))
  idx2 = sample((1:row_num)[-idx1], round(row_num/5))
  idx3 = sample((1:row_num)[-c(idx1, idx2)], round(row_num/5))
  idx4 = sample((1:row_num)[-c(idx1, idx2, idx3)], round(row_num/5))
  idx5 = (1:row_num)[-c(idx1, idx2, idx3, idx4)]
  idxV = list(idx1, idx2, idx3, idx4, idx5)
  for (i in 1:5){
    svmfit = e1071::svm(clusterRes ~ ., data = dataframe[-idxV[[i]],], kernel = "radial")
    pred[idxV[[i]]] = as.character(predict(svmfit, subset(dataframe, select = -clusterRes)[idxV[[i]],]))
    }
    conf.mat <- table(dataframe$clusterRes, as.factor(pred))
    simil_mat <- t(t(conf.mat)/apply(conf.mat,2,sum))
  return(simil_mat)
}
```

Start the testing

```{r}
countM = meta_Schiller2019 %>% 
  group_by(celltype, sample) %>% 
  summarize(n = n()) %>% 
  pivot_wider(names_from = celltype, values_from = n) %>% 
  filter(sample %in% c("old_1", "old_2", "young_7", "young_8")) %>% 
  select(-`NA`) %>% 
  column_to_rownames("sample")
countM[is.na(countM)] = 0
designM = data.frame(condition = c("old", "old", "young", "young"))
save(countM, designM, svm_mat, file = "/storage/holab/linxy/DCATS/real_world/Schiller_2019/dcats_input.RData")
```

```{r}
betabin_null = dcats_GLM(countM, designM)
betabin_null_ref = dcats_GLM(countM, designM, reference = "Club_cells")
wtoPhi_emSVM = dcats_GLM(countM, designM, similarity_mat = svm_mat)
estPhi = getPhi(countM, designM)
estPhi_null = dcats_GLM(countM, designM, fix_phi = estPhi)
estPhi_emSVM = dcats_GLM(countM, designM, similarity_mat = svm_mat, fix_phi = estPhi)
## Fisher's exact test
fisher_pvals = getFisher(as.matrix(countM[1:2,]), as.matrix(countM[3:4,]))

betabin_null = eachPhi(countM, designM)
wtoPhi_emSVM = eachPhi(countM, designM, similarity_mat = Haber2017$svm_mat)
estPhi = getPhi(countM, designM)
estPhi_null = dcats_GLM(countM, designM, fix_phi = estPhi)
estPhi_emSVM = dcats_GLM(countM, designM, similarity_mat = Haber2017$svm_mat, fix_phi = estPhi)
## Fisher's exact test
fisher_pvals = getFisher(as.matrix(Kang2017$count_ctrl), as.matrix(Kang2017$count_stim))
## speckle
speckleRes = speckle::propeller(clusters = exp6DF_mul$clusterRes, sample = exp6DF_mul$batch, group = exp6DF_mul$condition) %>% 
  dplyr::rename(cluster = BaselineProp.clusters,
                      speckle_pvals = FDR) %>%
  dplyr::select(cluster, speckle_pvals)
## scDC
cluster_num = dim(Kang2017$count_ctrl)[2]
res_scDC <- scDC_noClustering(cellTypes = exp6DF_mul$clusterRes, exp6DF_mul$batch, calCI = TRUE, calCI_method = c("percentile", "BCa", "multinom"),nboot = 1000, verbose = FALSE)
res_GLM <- fitGLM(res_scDC, c(rep("cond1",dim(Kang2017$count_ctrl)[1]*cluster_num),rep("cond2",dim(Kang2017$count_stim)[1]*cluster_num)), pairwise = FALSE, fixed_only = TRUE, verbose = FALSE)
scDCRes_temp = summary(res_GLM$pool_res_fixed)
scDCRes = scDCRes_temp[c(cluster_num+1,(dim(scDCRes_temp)[1]-cluster_num+2):dim(scDCRes_temp)[1]),]
```
```{r}
cbind(betabin_null$LRT_pvals < 0.05, betabin_null$fdr < 0.05, betabin_null_ref$LRT_pvals < 0.05, betabin_null_ref$fdr < 0.05)
```

